// @ruf/shell/esm2015/src/megamenu/search/search.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, ViewChild, ViewEncapsulation } from '@angular/core';
import { RufShellIntl } from '../../utils/shell-intl';
import { RufKeyCodes } from './../../input/keyboard';
let RufMegamenuSearchComponent = class RufMegamenuSearchComponent {
    constructor(_intl, changeDetectorRef) {
        this._intl = _intl;
        this.filterValue = new EventEmitter();
        this.inputFocus = new EventEmitter();
        this._intlChanges = _intl.changes.subscribe(() => changeDetectorRef.markForCheck());
    }
    get placeholder() {
        return this._placeholder || this._intl.megamenusearch.placeholder;
    }
    set placeholder(value) {
        this._placeholder = value;
    }
    get clearSearchLabel() {
        return this._clearSearchLabel || this._intl.megamenusearch.aria.clearSearchLabel;
    }
    set clearSearchLabel(value) {
        this._clearSearchLabel = value;
    }
    clear($event) {
        this.filter.nativeElement.value = '';
        this.filterValue.emit('');
        $event.preventDefault();
        $event.stopPropagation();
    }
    handleKeyDown($event) {
        switch ($event.keyCode) {
            case RufKeyCodes.SPACE:
            case RufKeyCodes.ENTER:
                this.clear($event);
                break;
        }
    }
    ngOnDestroy() {
        this._intlChanges.unsubscribe();
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufMegamenuSearchComponent.prototype, "placeholder", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufMegamenuSearchComponent.prototype, "clearSearchLabel", null);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], RufMegamenuSearchComponent.prototype, "filterValue", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], RufMegamenuSearchComponent.prototype, "inputFocus", void 0);
tslib_1.__decorate([
    ViewChild('filter', { static: false }),
    tslib_1.__metadata("design:type", ElementRef)
], RufMegamenuSearchComponent.prototype, "filter", void 0);
RufMegamenuSearchComponent = tslib_1.__decorate([
    Component({
        selector: 'ruf-megamenu-search',
        template: "<div class=\"fis-search-input-container\" [class.ruf-focus]='inputFocusClass'>\n  <div class=\"fis-icon-search\"></div>\n  <input type=\"text\" #filter (focus)=\"inputFocus.emit(filter.value);inputFocusClass = true\" (keyup.enter)=\"filterValue.emit(filter.value)\"\n         (blur)=\"inputFocusClass = false;\"\n         (input)=\"filterValue.emit(filter.value)\" [placeholder]=\"placeholder\"\n         [attr.aria-label]=\"placeholder\">\n  <i class=\"fis-icon-close\" (click)=\"clear($event);inputFocusClass = true;filter.focus();\"\n       (keydown)=\"handleKeyDown($event)\"\n       (blur)=\"inputFocusClass = false;\"\n       [title]=\"clearSearchLabel\"\n       role=\"button\" [attr.aria-label]=\"clearSearchLabel\" tabindex=\"0\"></i>\n</div>\n",
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: ["ruf-megamenu-search{padding-right:0}ruf-megamenu-search input{background-color:transparent;border:0 solid;content:'\\E93E';height:40px;outline:0;width:100%}ruf-megamenu-search input[type=text]::-ms-clear{display:none}ruf-megamenu-search .fis-icon-search{font-size:22px;margin-left:5px;margin-top:10px}ruf-megamenu-search .fis-icon-close{-ms-grid-row-align:center;align-self:center;height:1em;margin-right:5px;position:relative}ruf-megamenu-search .fis-search-input-container{display:-webkit-box;display:flex}"]
    }),
    tslib_1.__metadata("design:paramtypes", [RufShellIntl, ChangeDetectorRef])
], RufMegamenuSearchComponent);
export { RufMegamenuSearchComponent };
