// @ruf/navigation-accelerator/lib/files/template-app.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getAppComponentTs() {
    return `
import {
  Component,
  ViewEncapsulation,
  ViewChild,
  OnInit
} from '@angular/core';
import { Location } from '@angular/common';
import { RufDynamicSidemenuItem } from '@ruf/shell';
import { SidenavItems } from './sidenav-item'
import { Router } from '@angular/router';

@Component({
  selector: 'ruf48-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
  encapsulation: ViewEncapsulation.None

})
export class AppComponent implements OnInit {
  title = 'myapp';
  dir = 'ltr';
  @ViewChild('sidenav', { static: false }) sidenav;

  items: RufDynamicSidemenuItem[];

  constructor(
    private location: Location,
    private router: Router
  ) { }

  ngOnInit() {
    this.items = SidenavItems;
  }

  get selectedPath() {
    return this.location.path();
  }

  open(path: string) {
    if (this.matches(path)) {
      this.router.navigate([path]);
    }
    this.sidenav.close();
  }

  matches(path) {
    const pattern = new RegExp('(' + path + ')(\/.*)?$');
    if (pattern.test(this.location.path())) {
      return false;
    }
    return true;
  }
}

  `;
}
exports.getAppComponentTs = getAppComponentTs;
