// @ruf/navigation-accelerator/lib/files/template-app.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getAppModuleTs() {
    return `
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations'; // Don't like animations? Replace this with NoopAnimationsModule
import { NgModule } from '@angular/core';
import { RouterModule, RouteReuseStrategy } from '@angular/router';
import { MatIconModule, MatSidenavModule } from '@angular/material';
import { FlexLayoutModule } from '@angular/flex-layout';

import {
  RufAppCanvasModule,
  RufDynamicSideMenuModule,
  RufLayoutModule,
  RufBannerModule,
  RufFooterModule,
  RufIconModule,
  RufMenubarModule,
  RufNavbarModule,
  RufPageHeaderModule,
  RufMDIStrategy,
  RufRouterUtilService
} from '@ruf/shell';

import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';

@NgModule({
  declarations: [ AppComponent ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    RouterModule,
    FlexLayoutModule,
    MatIconModule,
    MatSidenavModule,
    BrowserAnimationsModule,
    RufAppCanvasModule,
    RufLayoutModule,
    RufBannerModule,
    RufDynamicSideMenuModule,
    RufFooterModule,
    RufIconModule,
    RufMenubarModule,
    RufNavbarModule,
    RufPageHeaderModule
  ],
  entryComponents: [],
  providers: [{
    provide: RouteReuseStrategy,
    useFactory: getMDIStrategy,
    deps: [RufRouterUtilService]
    },
    RufRouterUtilService
  ],
  bootstrap: [AppComponent]
})
export class AppModule {}

export class CustomMDIStrategy extends RufMDIStrategy {
  constructor(util: RufRouterUtilService) {
    super(util);
  }
}

export function getMDIStrategy(util: RufRouterUtilService) {
  return new CustomMDIStrategy(util);
}
`;
}
exports.getAppModuleTs = getAppModuleTs;
