// @ruf/shell/esm5/src/side-panel/side-panel.component.js
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation, ChangeDetectorRef, ChangeDetectionStrategy, Input, Output, EventEmitter } from '@angular/core';
import { RufShellIntl } from '../utils/shell-intl';
var RufSidePanelComponent = /** @class */ (function () {
    function RufSidePanelComponent(_intl, changeDetectorRef) {
        this._intl = _intl;
        this.changeDetectorRef = changeDetectorRef;
        /* tslint:disable:no-input-rename */
        this.isPinned = false;
        this.togglePin = new EventEmitter();
        this._intlChanges = _intl.changes.subscribe(function () {
            changeDetectorRef.markForCheck();
        });
    }
    Object.defineProperty(RufSidePanelComponent.prototype, "pinLabel", {
        get: function () {
            return this._pinLabel || this._intl.sidepanel.aria.pinLabel;
        },
        set: function (value) {
            this._pinLabel = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufSidePanelComponent.prototype, "unpinLabel", {
        get: function () {
            return this._unpinLabel || this._intl.sidepanel.aria.unpinLabel;
        },
        set: function (value) {
            this._unpinLabel = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufSidePanelComponent.prototype, "toggleLabel", {
        get: function () {
            return this.isPinned ? this.unpinLabel : this.pinLabel;
        },
        enumerable: true,
        configurable: true
    });
    RufSidePanelComponent.prototype.onPintoggle = function ($event) {
        this.isPinned = !this.isPinned;
        this.togglePin.emit({ 'pinned': this.isPinned });
    };
    RufSidePanelComponent.prototype.ngOnDestroy = function () {
        this._intlChanges.unsubscribe();
    };
    tslib_1.__decorate([
        Input('pinned'),
        tslib_1.__metadata("design:type", Object)
    ], RufSidePanelComponent.prototype, "isPinned", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], RufSidePanelComponent.prototype, "togglePin", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufSidePanelComponent.prototype, "pinLabel", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufSidePanelComponent.prototype, "unpinLabel", null);
    RufSidePanelComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-side-panel',
            template: "<div rufId class=\"ruf-side-panel-container\">\n  <button type=\"button\"\n          rufId=\"ruf_toggle_pin\"\n          class=\"ruf-side-panel-pin\"\n          mat-icon-button\n          fisStyle\n          [attr.aria-label]=\"toggleLabel\"\n          (click)=\"onPintoggle($event)\">\n      <mat-icon rufIconStyle=\"lg\" rufId=\"pin_icon\"\n              [fisIcon]=\"isPinned ? 'pin-remove': 'pin'\">\n      </mat-icon>\n  </button>\n  <ng-content></ng-content>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["ruf-side-panel .ruf-side-panel-container{border-style:solid;border-width:.0625rem;bottom:0;box-sizing:border-box;height:100%;max-width:25rem;min-width:18.75rem;overflow:auto;position:relative;top:0;width:100%}ruf-side-panel .ruf-side-panel-pin{float:right;margin:.3125rem}"]
        }),
        tslib_1.__metadata("design:paramtypes", [RufShellIntl,
            ChangeDetectorRef])
    ], RufSidePanelComponent);
    return RufSidePanelComponent;
}());
export { RufSidePanelComponent };
