// @ruf/bundler/lib/finder/finder.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
const template_1 = require("../template");
const path_1 = require("path");
const PACKAGE_FILE = 'package.json';
const TSCONFIG = 'tsconfig.json';
const EXAMPLE_DEFAULTS = {
    template: 'ruf-demo-workspace'
};
function findRootProjectDependencies(dir) {
    return __awaiter(this, void 0, void 0, function* () {
        const packageDir = yield util_1.findup(dir, TSCONFIG).catch(_ => '');
        if (packageDir) {
            const pkg = yield util_1.readJSONFile(path_1.join(packageDir, PACKAGE_FILE));
            return Object.assign({}, pkg.dependencies, pkg.peerDependencies, (pkg.rufExample && pkg.rufExample.dependencies));
        }
        return {};
    });
}
function findLibDependencies(dir) {
    return __awaiter(this, void 0, void 0, function* () {
        const packageDir = yield util_1.findup(dir, PACKAGE_FILE).catch(_ => '');
        if (packageDir.indexOf('/libs/') > 0) {
            const pkg = yield util_1.readJSONFile(path_1.join(packageDir, PACKAGE_FILE));
            return Object.assign({}, pkg.dependencies, pkg.peerDependencies, (pkg.rufExample && pkg.rufExample.dependencies), { [pkg.name]: pkg.version });
        }
        return {};
    });
}
function getTemplateDependencies(templateName) {
    const tpl = new template_1.Template(templateName);
    return Object.assign({}, tpl.dependencies);
}
const importRE = /import .+ from ['"]([@\w][^'"]+)['"]/gm;
const fromRE = /['"]([^'"]+)['"]/;
function findDepsInCode(files) {
    return [
        ...Object.keys(files).reduce((acc, cur) => {
            if (cur.endsWith('.ts')) {
                const imports = files[cur].match(importRE);
                if (imports) {
                    imports.map(m => m.match(fromRE)[1]).forEach(acc.add.bind(acc));
                }
            }
            return acc;
        }, new Set())
    ];
}
const rufCoreDeps = [
    '@ruf/fis-ux-assets',
    '@ruf/material',
    '@ruf/shell',
    '@ruf/pipes',
    '@ruf/theming'
];
const rufShell43Deps = {};
const rufMaterial43Deps = {};
const ruf43Deps = {
    '@ruf/shell': {
        '@angular/animations': '5.2.1',
        '@angular/cdk': '~5.2.3',
        '@angular/material': '~5.2.3',
        '@angular/common': '5.2.1',
        '@angular/compiler': '5.2.1',
        '@angular/core': '5.2.1',
        '@angular/flex-layout': '2.0.0-beta.12',
        '@angular/forms': '5.2.1',
        '@angular/http': '5.2.1',
        '@angular/platform-browser': '5.2.1',
        '@angular/platform-browser-dynamic': '5.2.1',
        '@angular/router': '5.2.1',
        '@angular/upgrade': '5.2.1',
        '@ruf/fis-ux-assets': '3.2.0',
        '@ruf/material': '5.0.2',
        '@ruf/theming': '4.4.0-beta1',
        'core-js': '2.5.4',
        rxjs: '5.5.8',
        tslib: '^1.7.1',
        'zone.js': '0.8.20'
    },
    '@ruf/material': {
        '@angular/cdk': '~5.2.3',
        '@angular/material': '~5.2.3'
    }
};
const re43 = /^[\^~]?4\.3\./;
function explodeRufDependencies(deps) {
    const exploded = Object.keys(deps)
        .filter(d => (d === '@ruf/shell' || d === '@ruf/material') && deps[d].match(re43))
        .map(d => ruf43Deps[d]);
    return Object.assign({}, deps, ...exploded);
}
function defineDependencies(rootDeps, dir, example) {
    return __awaiter(this, void 0, void 0, function* () {
        const depsPool = explodeRufDependencies(Object.assign({}, rootDeps, (yield findLibDependencies(dir)), example.dependencies));
        const required = new Set(Object.keys(explodeRufDependencies(getTemplateDependencies(example.template))).concat(findDepsInCode(example.files)));
        const diff = new Set(required);
        const pool = new Set(Object.keys(depsPool).concat(rufCoreDeps));
        for (const dep of required) {
            if (isDepPresent(dep, pool)) {
                diff.delete(dep);
            }
        }
        if (diff.size > 0) {
            throw new util_1.DetailedError(`Dependencies used in the code of '${example.title}' are missing.`, [...diff].toString());
        }
        return Object.assign({}, ...[...required].map(n => ({ [n]: depsPool[n] })));
    });
}
function isDepPresent(dep, pool) {
    if (pool.has(dep)) {
        return true;
    }
    if (dep.indexOf('/') > 1) {
        const parentDep = dep.substring(0, dep.lastIndexOf('/'));
        return isDepPresent(parentDep, pool);
    }
    return false;
}
function findExamples(opts, visit) {
    return __awaiter(this, void 0, void 0, function* () {
        opts.cwd = opts.cwd || process.cwd();
        const defaults = Object.assign({}, EXAMPLE_DEFAULTS, opts.defaults);
        const rootDependencies = yield findRootProjectDependencies(opts.cwd);
        const exampleFiles = yield util_1.glob('**/ruf-example.json', { cwd: opts.cwd });
        for (const file of exampleFiles) {
            const path = path_1.join(opts.cwd, file);
            let example = JSON.parse((yield util_1.readFile(path)).toString());
            const exampleDir = path_1.dirname(path);
            example.dependencies = Object.assign({}, defaults.dependencies, example.dependencies);
            example = Object.assign({}, defaults, example);
            const loadedFiles = yield util_1.getLoadedFiles({
                cwd: exampleDir,
                assets: opts.assets
            });
            example.files = Object.assign({}, example.files, loadedFiles);
            example.dependencies = yield defineDependencies(rootDependencies, exampleDir, example);
            delete example.ignore;
            visit(example, { dir: exampleDir });
        }
    });
}
exports.findExamples = findExamples;
function findReadmes(opts, visit) {
    return __awaiter(this, void 0, void 0, function* () {
        opts.cwd = opts.cwd || process.cwd();
        const readmeFiles = yield util_1.glob('**/README.md', { cwd: opts.cwd });
        for (const file of readmeFiles) {
            const path = path_1.join(opts.cwd, file);
            const loadedFile = (yield util_1.readFile(path)).toString();
            const readmeFile = { prop: path, data: loadedFile };
            visit(readmeFile);
        }
    });
}
exports.findReadmes = findReadmes;
