// @ruf/shell/esm2015/src/toggle/toggle.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostBinding, HostListener, Input, Output, EventEmitter } from '@angular/core';
import { ENTER, SPACE } from '@angular/cdk/keycodes';
import { CdkOverlayOrigin } from '@angular/cdk/overlay';
/**
 * Directive that can be applied on any element, and that toggles an attribute that can be used by the underlying element as need be.
 */
let RufToggleDirective = class RufToggleDirective extends CdkOverlayOrigin {
    constructor(elRef) {
        super(elRef);
        this.elRef = elRef;
        // boolean property that gets toggled when the underlying element is clicked
        // tslint:disable-next-line:no-input-rename
        this.value = false;
        this.chevron = false;
        this.arrow = false;
        this.toggleChange = new EventEmitter();
        this.toggle = true;
    }
    get active() {
        return this.value;
    }
    set rufToggle(val) {
        this.value = val || false;
    }
    set rufToggleChevron(val) {
        this.arrow = false;
        this.chevron = true;
        this.value = val || false;
    }
    set rufToggleArrow(val) {
        this.arrow = true;
        this.chevron = false;
        this.value = val || false;
    }
    get normal() {
        return !this.chevron && !this.arrow;
    }
    // select should focus the element and toggle value
    select(event) {
        this.elRef.nativeElement.focus();
        this.toggleValue(event);
    }
    toggleValue(event) {
        event.stopPropagation();
        event.preventDefault();
        // For Safari browser, event.stopPropagation() is not setting the value of
        // event.cancelBubble as true. So, we need to explicitly set it.
        event.cancelBubble = true;
        this.value = !this.value;
        this.toggleChange.emit({ trigger: this.elRef, value: this.value, event: event });
    }
    onClick(event) {
        this.toggleValue(event);
    }
    onKey(event) {
        switch (event.keyCode) {
            case ENTER:
            case SPACE:
                this.toggleValue(event);
                break;
            default:
        }
    }
};
tslib_1.__decorate([
    HostBinding('class.ruf-toggle-active'),
    tslib_1.__metadata("design:type", Object)
], RufToggleDirective.prototype, "value", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-toggle-chevron'),
    tslib_1.__metadata("design:type", Object)
], RufToggleDirective.prototype, "chevron", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-toggle-arrow'),
    tslib_1.__metadata("design:type", Object)
], RufToggleDirective.prototype, "arrow", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-active'),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [])
], RufToggleDirective.prototype, "active", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], RufToggleDirective.prototype, "rufToggle", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], RufToggleDirective.prototype, "rufToggleChevron", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], RufToggleDirective.prototype, "rufToggleArrow", null);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], RufToggleDirective.prototype, "toggleChange", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-toggle'),
    tslib_1.__metadata("design:type", Object)
], RufToggleDirective.prototype, "toggle", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-toggle-normal'),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [])
], RufToggleDirective.prototype, "normal", null);
tslib_1.__decorate([
    HostListener('click', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], RufToggleDirective.prototype, "onClick", null);
tslib_1.__decorate([
    HostListener('keydown', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [KeyboardEvent]),
    tslib_1.__metadata("design:returntype", void 0)
], RufToggleDirective.prototype, "onKey", null);
RufToggleDirective = tslib_1.__decorate([
    Directive({
        selector: '[rufToggle], [rufToggleChevron], [rufToggleArrow]',
        exportAs: 'rufToggle'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef])
], RufToggleDirective);
export { RufToggleDirective };
