// @ruf/shell/esm2015/src/router/outlet.directive.js
import * as tslib_1 from "tslib";
import { Attribute, ComponentFactoryResolver, ChangeDetectorRef, Directive, ViewContainerRef } from '@angular/core';
import { RouterOutlet, ChildrenOutletContexts, PRIMARY_OUTLET } from '@angular/router';
import { RufRouterUtilService } from './router-util.service';
let RufRouterOutletDirective = class RufRouterOutletDirective extends RouterOutlet {
    constructor(rufParentContexts, rufLocation, rufResolver, rufName, rufChangeDetector, util) {
        super(rufParentContexts, rufLocation, rufResolver, rufName, rufChangeDetector);
        this.rufParentContexts = rufParentContexts;
        this.rufLocation = rufLocation;
        this.rufResolver = rufResolver;
        this.rufName = rufName;
        this.rufChangeDetector = rufChangeDetector;
        this.util = util;
        this.rufName = rufName || PRIMARY_OUTLET;
    }
    detach() {
        const componentRef = super.detach();
        const lastActivatedOutlet = this.util.lastActivatedOutlet;
        // Detach child routes when a parent route like /banner or /layout is being detached.
        if (lastActivatedOutlet && lastActivatedOutlet.isActivated) {
            const lastCompRef = lastActivatedOutlet.detach();
            this.util.setComponentRef(this.util.lastActivatedPath, lastCompRef);
        }
        else {
            this.util.setComponentRef(this.util.lastActivatedPath, componentRef);
        }
        return componentRef;
    }
    activateWith(activatedRoute, resolver) {
        this.storePreviousRoute(activatedRoute);
        super.activateWith(activatedRoute, resolver);
    }
    attach(ref, activatedRoute) {
        this.storePreviousRoute(activatedRoute);
        super.attach(ref, activatedRoute);
    }
    storePreviousRoute(activatedRoute) {
        this.util.lastActivatedPath = this.util.fullPath(activatedRoute.snapshot);
        if (!activatedRoute.children || activatedRoute.children.length === 0) {
            // Store router outlet only in case of leaf routes. This will be used when detaching parent routes.
            this.util.lastActivatedOutlet = this.rufParentContexts.getContext(this.rufName).outlet;
        }
    }
};
RufRouterOutletDirective = tslib_1.__decorate([
    Directive({
        selector: 'ruf-router-outlet' // tslint:disable-line
    }),
    tslib_1.__param(3, Attribute('name')),
    tslib_1.__metadata("design:paramtypes", [ChildrenOutletContexts,
        ViewContainerRef,
        ComponentFactoryResolver, String, ChangeDetectorRef,
        RufRouterUtilService])
], RufRouterOutletDirective);
export { RufRouterOutletDirective };
