// @ruf/shell/esm2015/src/router/router-util.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
let RufRouterUtilService = class RufRouterUtilService {
    constructor() {
        this._routeComp = {};
    }
    set lastActivatedPath(path) {
        this._lastActivatedPath = path;
    }
    get lastActivatedPath() {
        return this._lastActivatedPath;
    }
    set lastActivatedOutlet(outlet) {
        this._lastActivatedOutlet = outlet;
    }
    get lastActivatedOutlet() {
        return this._lastActivatedOutlet;
    }
    setComponentRef(path, componentRef) {
        this._routeComp[path] = componentRef;
    }
    getComponentRef(path) {
        return this._routeComp[path];
    }
    fullPath(snapshot) {
        let path = '';
        const rootElems = snapshot.pathFromRoot;
        rootElems.forEach(function (child) {
            const urlArray = child.url;
            urlArray.forEach(function (url) {
                path += `/${url.path}`;
            });
        });
        return path;
    }
    getLeafNodePath(route) {
        let path;
        // if (!route.routeConfig.loadChildren) {
        // If !route.component, this is a parent route. Return path of the leaf node in this case.
        if (!route.component) {
            const leafNode = this.getLeafNode(route);
            path = this.fullPath(leafNode);
        }
        else {
            path = this.fullPath(route);
        }
        // }
        return path;
    }
    getLeafNode(route) {
        if (!route.children || route.children.length === 0) {
            return route;
        }
        return this.getLeafNode(route.firstChild || route.children[0]);
    }
};
RufRouterUtilService = tslib_1.__decorate([
    Injectable()
], RufRouterUtilService);
export { RufRouterUtilService };
