// @ruf/shell/esm5/src/layout/margin.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostBinding, Input, Renderer2, ElementRef } from '@angular/core';
export var RUF_MARGIN_VALUES = ['none', 'auto', 'small', 'medium', 'large'];
var RufMarginDirective = /** @class */ (function () {
    function RufMarginDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.classPrefix = 'ruf-margin';
        this._positions = {
            rufMargin: ['-'],
            rufMarginVertical: ['-top-', '-bottom-'],
            rufMarginHorizontal: ['-left-', '-right-'],
            rufMarginTop: ['-top-'],
            rufMarginRight: ['-right-'],
            rufMarginBottom: ['-bottom-'],
            rufMarginLeft: ['-left-']
        };
        this.rufMargin = '';
        this.rufMarginVertical = '';
        this.rufMarginHorizontal = '';
        this.rufMarginTop = '';
        this.rufMarginRight = '';
        this.rufMarginBottom = '';
        this.rufMarginLeft = '';
        this.margin = true;
    }
    RufMarginDirective.prototype.ngOnChanges = function (changes) {
        var _this = this;
        Object.keys(changes).forEach(function (key) {
            _this.update(key, changes[key]);
        });
    };
    RufMarginDirective.prototype.update = function (key, change) {
        var _this = this;
        var previous = this._validateValue(change.previousValue);
        var current = this._validateValue(change.currentValue);
        var positions = this._positions[key];
        positions.forEach(function (pos) {
            // remove
            _this.renderer.removeClass(_this.el.nativeElement, _this.classPrefix + pos + previous);
            // add
            _this.renderer.addClass(_this.el.nativeElement, _this.classPrefix + pos + current);
        });
    };
    RufMarginDirective.prototype._validateValue = function (val) {
        if (RUF_MARGIN_VALUES.indexOf(val) !== -1) {
            return val;
        }
        return 'medium';
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "rufMargin", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "rufMarginVertical", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "rufMarginHorizontal", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "rufMarginTop", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "rufMarginRight", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "rufMarginBottom", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "rufMarginLeft", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-margin'),
        tslib_1.__metadata("design:type", Object)
    ], RufMarginDirective.prototype, "margin", void 0);
    RufMarginDirective = tslib_1.__decorate([
        Directive({
            selector: "\n    [rufMargin],\n    [rufMarginVertical],\n    [rufMarginHorizontal],\n    [rufMarginTop],\n    [rufMarginRight],\n    [rufMarginBottom],\n    [rufMarginLeft]\n  ",
            exportAs: 'rufMargin'
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], RufMarginDirective);
    return RufMarginDirective;
}());
export { RufMarginDirective };
