// @ruf/shell/esm5/src/layout/padding.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostBinding, Input, Renderer2, ElementRef } from '@angular/core';
export var RUF_PADDING_VALUES = ['none', 'auto', 'small', 'medium', 'large'];
var RufPaddingDirective = /** @class */ (function () {
    function RufPaddingDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.classPrefix = 'ruf-padding';
        this._positions = {
            rufPadding: ['-'],
            rufPaddingVertical: ['-top-', '-bottom-'],
            rufPaddingHorizontal: ['-left-', '-right-'],
            rufPaddingTop: ['-top-'],
            rufPaddingRight: ['-right-'],
            rufPaddingBottom: ['-bottom-'],
            rufPaddingLeft: ['-left-']
        };
        this.rufPadding = '';
        this.rufPaddingVertical = '';
        this.rufPaddingHorizontal = '';
        this.rufPaddingTop = '';
        this.rufPaddingRight = '';
        this.rufPaddingBottom = '';
        this.rufPaddingLeft = '';
        this.padding = true;
    }
    RufPaddingDirective.prototype.ngOnChanges = function (changes) {
        var _this = this;
        Object.keys(changes).forEach(function (key) {
            _this.update(key, changes[key]);
        });
    };
    RufPaddingDirective.prototype.update = function (key, change) {
        var _this = this;
        var previous = this._validateValue(change.previousValue);
        var current = this._validateValue(change.currentValue);
        var positions = this._positions[key];
        positions.forEach(function (pos) {
            // remove
            _this.renderer.removeClass(_this.el.nativeElement, _this.classPrefix + pos + previous);
            // add
            _this.renderer.addClass(_this.el.nativeElement, _this.classPrefix + pos + current);
        });
    };
    RufPaddingDirective.prototype._validateValue = function (val) {
        if (RUF_PADDING_VALUES.indexOf(val) !== -1) {
            return val;
        }
        return 'medium';
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "rufPadding", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "rufPaddingVertical", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "rufPaddingHorizontal", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "rufPaddingTop", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "rufPaddingRight", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "rufPaddingBottom", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "rufPaddingLeft", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-padding'),
        tslib_1.__metadata("design:type", Object)
    ], RufPaddingDirective.prototype, "padding", void 0);
    RufPaddingDirective = tslib_1.__decorate([
        Directive({
            selector: "\n    [rufPadding],\n    [rufPaddingVertical],\n    [rufPaddingHorizontal],\n    [rufPaddingTop],\n    [rufPaddingRight],\n    [rufPaddingBottom],\n    [rufPaddingLeft]\n  ",
            exportAs: 'rufPadding'
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], RufPaddingDirective);
    return RufPaddingDirective;
}());
export { RufPaddingDirective };
