// @ruf/shell/esm2015/src/layout/margin.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostBinding, Input, Renderer2, ElementRef } from '@angular/core';
export const RUF_MARGIN_VALUES = ['none', 'auto', 'small', 'medium', 'large'];
let RufMarginDirective = class RufMarginDirective {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.classPrefix = 'ruf-margin';
        this._positions = {
            rufMargin: ['-'],
            rufMarginVertical: ['-top-', '-bottom-'],
            rufMarginHorizontal: ['-left-', '-right-'],
            rufMarginTop: ['-top-'],
            rufMarginRight: ['-right-'],
            rufMarginBottom: ['-bottom-'],
            rufMarginLeft: ['-left-']
        };
        this.rufMargin = '';
        this.rufMarginVertical = '';
        this.rufMarginHorizontal = '';
        this.rufMarginTop = '';
        this.rufMarginRight = '';
        this.rufMarginBottom = '';
        this.rufMarginLeft = '';
        this.margin = true;
    }
    ngOnChanges(changes) {
        Object.keys(changes).forEach(key => {
            this.update(key, changes[key]);
        });
    }
    update(key, change) {
        const previous = this._validateValue(change.previousValue);
        const current = this._validateValue(change.currentValue);
        const positions = this._positions[key];
        positions.forEach(pos => {
            // remove
            this.renderer.removeClass(this.el.nativeElement, this.classPrefix + pos + previous);
            // add
            this.renderer.addClass(this.el.nativeElement, this.classPrefix + pos + current);
        });
    }
    _validateValue(val) {
        if (RUF_MARGIN_VALUES.indexOf(val) !== -1) {
            return val;
        }
        return 'medium';
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "rufMargin", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "rufMarginVertical", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "rufMarginHorizontal", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "rufMarginTop", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "rufMarginRight", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "rufMarginBottom", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "rufMarginLeft", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-margin'),
    tslib_1.__metadata("design:type", Object)
], RufMarginDirective.prototype, "margin", void 0);
RufMarginDirective = tslib_1.__decorate([
    Directive({
        selector: `
    [rufMargin],
    [rufMarginVertical],
    [rufMarginHorizontal],
    [rufMarginTop],
    [rufMarginRight],
    [rufMarginBottom],
    [rufMarginLeft]
  `,
        exportAs: 'rufMargin'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], RufMarginDirective);
export { RufMarginDirective };
