// @ruf/shell/esm2015/src/layout/padding.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostBinding, Input, Renderer2, ElementRef } from '@angular/core';
export const RUF_PADDING_VALUES = ['none', 'auto', 'small', 'medium', 'large'];
let RufPaddingDirective = class RufPaddingDirective {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.classPrefix = 'ruf-padding';
        this._positions = {
            rufPadding: ['-'],
            rufPaddingVertical: ['-top-', '-bottom-'],
            rufPaddingHorizontal: ['-left-', '-right-'],
            rufPaddingTop: ['-top-'],
            rufPaddingRight: ['-right-'],
            rufPaddingBottom: ['-bottom-'],
            rufPaddingLeft: ['-left-']
        };
        this.rufPadding = '';
        this.rufPaddingVertical = '';
        this.rufPaddingHorizontal = '';
        this.rufPaddingTop = '';
        this.rufPaddingRight = '';
        this.rufPaddingBottom = '';
        this.rufPaddingLeft = '';
        this.padding = true;
    }
    ngOnChanges(changes) {
        Object.keys(changes).forEach(key => {
            this.update(key, changes[key]);
        });
    }
    update(key, change) {
        const previous = this._validateValue(change.previousValue);
        const current = this._validateValue(change.currentValue);
        const positions = this._positions[key];
        positions.forEach(pos => {
            // remove
            this.renderer.removeClass(this.el.nativeElement, this.classPrefix + pos + previous);
            // add
            this.renderer.addClass(this.el.nativeElement, this.classPrefix + pos + current);
        });
    }
    _validateValue(val) {
        if (RUF_PADDING_VALUES.indexOf(val) !== -1) {
            return val;
        }
        return 'medium';
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "rufPadding", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "rufPaddingVertical", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "rufPaddingHorizontal", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "rufPaddingTop", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "rufPaddingRight", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "rufPaddingBottom", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "rufPaddingLeft", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-padding'),
    tslib_1.__metadata("design:type", Object)
], RufPaddingDirective.prototype, "padding", void 0);
RufPaddingDirective = tslib_1.__decorate([
    Directive({
        selector: `
    [rufPadding],
    [rufPaddingVertical],
    [rufPaddingHorizontal],
    [rufPaddingTop],
    [rufPaddingRight],
    [rufPaddingBottom],
    [rufPaddingLeft]
  `,
        exportAs: 'rufPadding'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], RufPaddingDirective);
export { RufPaddingDirective };
