// @ruf/fis-mega-menu/karma.conf.js
// Karma configuration
// Generated on Thu Oct 03 2013 14:05:38 GMT+0200 (SAST)

module.exports = function(config) {
    config.set({
        // base path, that will be used to resolve files and exclude
        basePath: '',
        frameworks: ['jasmine'],
        // list of files / patterns to load in the browser
        files: [
            // bower:js
            "demo/vendor/jquery/dist/jquery.js",
            "demo/vendor/get-size/get-size.js",
            "demo/vendor/ev-emitter/ev-emitter.js",
            "demo/vendor/desandro-matches-selector/matches-selector.js",
            "demo/vendor/fizzy-ui-utils/utils.js",
            "demo/vendor/outlayer/item.js",
            "demo/vendor/outlayer/outlayer.js",
            "demo/vendor/masonry/masonry.js",
            "demo/vendor/angular/angular.js",
            "demo/vendor/underscore/underscore.js",
            "demo/vendor/angular-resource/angular-resource.js",
            "demo/vendor/angular-route/angular-route.js",
            "demo/vendor/angular-animate/angular-animate.js",
            "demo/vendor/angular-sanitize/angular-sanitize.js",
            "demo/vendor/kendo-ui/src/js/kendo.all.js",
            "demo/vendor/moment/moment.js",
            "demo/vendor/keymaster/keymaster.js",
            "demo/vendor/angular-bootstrap/ui-bootstrap-tpls.js",
            "demo/vendor/fis-bootstrap-components/dist/fis-bootstrap-components.js",
            "demo/vendor/angular-smart-table/dist/smart-table.js",
            "demo/vendor/fis-bootstrap-table/dist/fis-bootstrap-table.js",
            "demo/vendor/fis-framework/fis-fmk-core.js",
            "demo/vendor/bootstrap/dist/js/bootstrap.js",
            "demo/vendor/angular-mocks/angular-mocks.js",
            // endbower
            //Add angular-mocks dependency
            'demo/vendor/angular-mocks/angular-mocks.js',
            // js-dependencies
            'src/scripts/**/*.module.js',
            'src/scripts/**/*.js',

            // do not change the order of the spec files
            // the fisMegaMenuDirectiveSpec file does not work if the service spec file is not executed first
            'test/spec/fisMegaMenuSearchServiceSpec.js',
            'test/spec/fisMegaMenuDirectiveSpec.js',
            'test/spec/**/*Spec.js'
        ],

        // list of files to exclude
        exclude: [],

        // test results reporter to use
        // possible values: dots || progress || growl
        reporters: ['dots'],

        // web server port
        port: 8080,

        // cli runner port
        runnerPort: 9100,

        // enable / disable colors in the output (reporters and logs)
        colors: true,

        // level of logging
        // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
        logLevel: config.LOG_INFO,

        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: false,

        // Start these browsers, currently available:
        // - Chrome
        // - ChromeCanary
        // - Firefox
        // - Opera
        // - Safari (only Mac)
        // - PhantomJS
        // - IE (only Windows)
        browsers: ['Chrome'],

        // If browser does not capture in given timeout [ms], kill it
        captureTimeout: 10000,

        // Continuous Integration mode
        // if true, it capture browsers, run tests and exit
        singleRun: true
    });
};
