// @ruf/shell/esm2015/src/unique-id/unique-id.directive.js
import * as tslib_1 from "tslib";
/* tslint:disable:no-bitwise */
/* tslint:disable:directive-selector-name */
import { Directive, ElementRef, isDevMode, Input } from '@angular/core';
let RufUniqueIdDirective = class RufUniqueIdDirective {
    constructor(el) {
        this.el = el;
    }
    ngOnInit() {
        if (isDevMode()) {
            const currentElement = this.el.nativeElement;
            if (!currentElement.id) {
                currentElement.id = this.rufId ? this.getUniqueId(this.rufId) : this.generateID();
            }
        }
    }
    /**
     * Walks up the tree, picking up uids as it goes
     * @param node
     * @returns parent ID
     */
    getParentID(node) {
        let tempParentId = '';
        if (node.attributes && node.attributes.id) {
            return node.attributes.id.value;
        }
        if (node.parentNode) {
            tempParentId = this.getParentID(node.parentNode);
        }
        return tempParentId;
    }
    /**
     * Generates an idempotent id on the element this directive is attached to
     * Based on the name and attributes (but ignores the angular temp name that can be different each run)
     * @returns string
     */
    generateID() {
        const nel = this.el.nativeElement;
        const parentID = this.getParentID(nel);
        let tempUid = parentID === '' ? nel.tagName.toLowerCase() : parentID + '_' + nel.nodeName.toLowerCase();
        tempUid = this.getUniqueId(tempUid);
        return tempUid;
    }
    /**
     * Check for unique id in DOM
     * @returns string
     */
    getUniqueId(tempId) {
        let count = 0;
        const genId = tempId + '_';
        tempId = genId + count;
        while (document.getElementById(tempId) !== null) {
            count++;
            tempId = genId + count;
        }
        return tempId;
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], RufUniqueIdDirective.prototype, "rufId", void 0);
RufUniqueIdDirective = tslib_1.__decorate([
    Directive({
        selector: '[rufId]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef])
], RufUniqueIdDirective);
export { RufUniqueIdDirective };
