// @ruf/shell/esm5/src/page-header/page-header.component.js
import * as tslib_1 from "tslib";
import { Component, Directive, ViewEncapsulation, Input } from '@angular/core';
var RufPageHeaderRowDirective = /** @class */ (function () {
    function RufPageHeaderRowDirective() {
    }
    RufPageHeaderRowDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-page-header-row' // tslint:disable-line
        })
    ], RufPageHeaderRowDirective);
    return RufPageHeaderRowDirective;
}());
export { RufPageHeaderRowDirective };
var RufPageHeaderIconComponent = /** @class */ (function () {
    function RufPageHeaderIconComponent() {
    }
    RufPageHeaderIconComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-page-header-icon',
            encapsulation: ViewEncapsulation.None,
            template: "\n    <div><ng-content></ng-content></div>\n  "
        })
    ], RufPageHeaderIconComponent);
    return RufPageHeaderIconComponent;
}());
export { RufPageHeaderIconComponent };
var RufPageHeaderTitleComponent = /** @class */ (function () {
    function RufPageHeaderTitleComponent() {
    }
    RufPageHeaderTitleComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-page-header-title',
            encapsulation: ViewEncapsulation.None,
            template: "\n    <ng-content></ng-content>\n  "
        })
    ], RufPageHeaderTitleComponent);
    return RufPageHeaderTitleComponent;
}());
export { RufPageHeaderTitleComponent };
var RufPageHeaderDescriptionComponent = /** @class */ (function () {
    function RufPageHeaderDescriptionComponent() {
    }
    RufPageHeaderDescriptionComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-page-header-description',
            encapsulation: ViewEncapsulation.None,
            template: "\n    <ng-content></ng-content>\n  "
        })
    ], RufPageHeaderDescriptionComponent);
    return RufPageHeaderDescriptionComponent;
}());
export { RufPageHeaderDescriptionComponent };
var RufPageHeaderComponent = /** @class */ (function () {
    function RufPageHeaderComponent() {
    }
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufPageHeaderComponent.prototype, "name", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufPageHeaderComponent.prototype, "title", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufPageHeaderComponent.prototype, "description", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufPageHeaderComponent.prototype, "icon", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufPageHeaderComponent.prototype, "color", void 0);
    RufPageHeaderComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-page-header',
            template: "<ruf-toolbar [color]=\"color\" rufId=\"pageheader_toolbar\" class=\"ruf-pageheader-toolbar\" >\n  <ruf-page-header-icon rufId=\"pageheader_icon\" *ngIf=\"icon\" class=\"ruf-page-header-text-icon\">{{icon}}</ruf-page-header-icon>\n  <ng-content select=\"ruf-page-header-icon\"></ng-content>\n  <div class=\"ruf-page-header-title-area\">\n    <ruf-page-header-title *ngIf=\"title\" rufId=\"pageheader_title\">{{title}}</ruf-page-header-title>\n    <ng-content rufId=\"pageheader_title_tag\" select=\"ruf-page-header-title\"></ng-content>\n    <ruf-page-header-description rufId=\"pageheader_description\" *ngIf=\"description\">{{description}}</ruf-page-header-description>\n    <ng-content rufId=\"pageheader_title_tag\" select=\"ruf-page-header-description\"></ng-content>\n  </div>\n  <div rufToolbarFiller></div>\n  <ng-content></ng-content>\n  <ng-content select=\"ruf-page-header-row\" ngProjectAs=\"ruf-toolbar-row\"></ng-content>\n</ruf-toolbar>\n",
            encapsulation: ViewEncapsulation.None,
            styles: ["ruf-page-header{display:block;min-height:2.375rem;width:100%}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows>.ruf-toolbar-first-row{-webkit-box-align:end;align-items:flex-end;padding:.625rem 0}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows .ruf-page-header-title-area{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;-webkit-box-pack:center;justify-content:center;min-height:2.5rem}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows .ruf-page-header-metric,ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows ruf-page-header-title{margin:0;text-transform:capitalize}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows .ruf-page-header-metric-description,ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows ruf-page-header-description{text-transform:initial}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows>ruf-page-header-row,ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows>ruf-toolbar-row{margin:0 auto;padding:0;width:100%}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows ruf-page-header-icon{border-radius:50%;flex-shrink:0;height:2.5rem;overflow:hidden;text-align:center;width:2.5rem}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows ruf-page-header-icon>div{font-size:1.875rem;line-height:2.625rem}ruf-page-header .ruf-pageheader-toolbar>.ruf-toolbar-rows ruf-page-header-icon.ruf-page-header-text-icon>div{font-size:1.125rem;line-height:2.5rem;text-transform:lowercase}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], RufPageHeaderComponent);
    return RufPageHeaderComponent;
}());
export { RufPageHeaderComponent };
