// @ruf/shell/esm5/src/navbar/navbar.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewEncapsulation, HostBinding } from '@angular/core';
var RufNavbarComponent = /** @class */ (function () {
    function RufNavbarComponent() {
        this.role = 'navigation';
    }
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufNavbarComponent.prototype, "color", void 0);
    tslib_1.__decorate([
        HostBinding('attr.role'),
        tslib_1.__metadata("design:type", Object)
    ], RufNavbarComponent.prototype, "role", void 0);
    RufNavbarComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-navbar',
            template: "<ruf-toolbar [color]=\"color\" rufId=\"navbar_toolbar\">\n  <ng-content></ng-content>\n</ruf-toolbar>\n",
            encapsulation: ViewEncapsulation.None,
            styles: ["ruf-navbar{display:block;flex-shrink:0;min-height:2.8125rem;overflow:hidden}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], RufNavbarComponent);
    return RufNavbarComponent;
}());
export { RufNavbarComponent };
