// @ruf/shell/esm5/src/sidemenu/sidemenu-children.component.js
import * as tslib_1 from "tslib";
import { trigger, state, style, transition, animate } from '@angular/animations';
import { Component, HostBinding, ViewEncapsulation, Input, ContentChildren, QueryList, forwardRef, ElementRef, HostListener } from '@angular/core';
import { RufSidemenuItemComponent } from './sidemenu-item.component';
import { RufKeyboardNavigation, RufKeyCodes } from '../input/keyboard';
import { RufSidemenuComponent } from './sidemenu.component';
import { RufToggleDirective } from '../toggle/toggle.directive';
/**
 * description
 */
var RufSidemenuChildrenComponent = /** @class */ (function () {
    function RufSidemenuChildrenComponent(_elementRef, sidemenuComp) {
        this._elementRef = _elementRef;
        this.sidemenuComp = sidemenuComp;
        this.visible = true;
        this.sidemenuChildren = true;
    }
    RufSidemenuChildrenComponent.prototype.ngOnChanges = function (changes) {
        if (changes.visible) {
            this.visibleState = changes.visible.currentValue;
        }
    };
    RufSidemenuChildrenComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (this.childItems && this.childItems.length > 0) {
            this._keyboardSideNav = new RufKeyboardNavigation();
            this._keyboardSideNav.init(this._elementRef, this.childItems).subscribe(function (path) {
                _this.sidemenuComp.select(path.path, path.data);
            });
        }
    };
    RufSidemenuChildrenComponent.prototype.setActiveItem = function (event) {
        // set active item when parent is clicked
        if (event.toState) {
            this._keyboardSideNav.setActiveItem(0);
        }
    };
    RufSidemenuChildrenComponent.prototype.onKey = function (event) {
        switch (event.keyCode) {
            case RufKeyCodes.ESCAPE:
                if (this.menuTrigger) {
                    this.menuTrigger.select(event);
                }
                break;
            default:
                if (this._keyboardSideNav && this._keyboardSideNav.isActive()) {
                    this._keyboardSideNav.onKeyDown(event);
                }
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufSidemenuChildrenComponent.prototype, "visible", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", RufToggleDirective)
    ], RufSidemenuChildrenComponent.prototype, "menuTrigger", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-sidemenu-children'),
        tslib_1.__metadata("design:type", Object)
    ], RufSidemenuChildrenComponent.prototype, "sidemenuChildren", void 0);
    tslib_1.__decorate([
        HostBinding('@visible'),
        tslib_1.__metadata("design:type", Boolean)
    ], RufSidemenuChildrenComponent.prototype, "visibleState", void 0);
    tslib_1.__decorate([
        ContentChildren(forwardRef(function () { return RufSidemenuItemComponent; })),
        tslib_1.__metadata("design:type", QueryList)
    ], RufSidemenuChildrenComponent.prototype, "childItems", void 0);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [KeyboardEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufSidemenuChildrenComponent.prototype, "onKey", null);
    RufSidemenuChildrenComponent = tslib_1.__decorate([
        Component({
            selector: "[rufSidemenuChildren], ruf-sidemenu-children",
            exportAs: 'rufSidemenuChildren',
            template: "<div (@visible.done)=\"setActiveItem($event)\" [@visible]=\"visible\"><ng-content></ng-content></div>",
            encapsulation: ViewEncapsulation.None,
            animations: [
                trigger('visible', [
                    state('0', style({
                        height: 0
                    })),
                    state('1', style({
                        height: '*'
                    })),
                    transition('0 => 1', animate('300ms ease-in-out')),
                    transition('1 => 0', animate('300ms ease-in-out'))
                ])
            ]
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, RufSidemenuComponent])
    ], RufSidemenuChildrenComponent);
    return RufSidemenuChildrenComponent;
}());
export { RufSidemenuChildrenComponent };
