// @ruf/shell/esm2015/src/utils/shell-intl.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * Labels used by all shell components that require internationalization.
 * These components also have individual attributes for these labels.
 * Users can either provide attributes
 * e.g., `closeIconTitle` can be provided by users for title of modal dialog close button.
 * <button mat-icon-button mat-dialog-close [title]="closeIconTitle">
 * </button>
 * or
 * Provide an instance of this class which overrides `closeIconTitle`.
 * */
let RufShellIntl = class RufShellIntl {
    /**
     * Labels used by all shell components that require internationalization.
     * These components also have individual attributes for these labels.
     * Users can either provide attributes
     * e.g., `closeIconTitle` can be provided by users for title of modal dialog close button.
     * <button mat-icon-button mat-dialog-close [title]="closeIconTitle">
     * </button>
     * or
     * Provide an instance of this class which overrides `closeIconTitle`.
     * */
    constructor() {
        /**
         * Stream that emits whenever the labels here are changed. Use this to notify
         * components if the labels have changed after initialization.
         */
        this.changes = new Subject();
        this.dynamicmenubar = {
            aria: { closeIconLabel: 'Close' }
        };
        this.dialog = { closeIconTitle: 'Close' };
        this.banner = {
            altText: 'logo',
            aria: {
                label: 'Banner',
                labelledBy: null
            }
        };
        this.toolbar = {
            aria: { label: 'Toolbar' }
        };
        this.statusbar = {
            aria: { closeIconLabel: 'Close', expandIconLabel: 'Expand' }
        };
        this.ellipsis = {
            label: '...Read More',
            aria: { label: 'Read More' }
        };
        this.megamenu = {
            placeholder: 'Search',
            aria: { clearSearchLabel: 'Clear search' }
        };
        this.megamenusearch = {
            placeholder: 'Search',
            aria: { clearSearchLabel: 'Clear search' }
        };
        this.card = {
            aria: {
                expandLabel: 'Expand',
                collapseLabel: 'Collapse'
            }
        };
        this.fileupload = {
            aria: {
                label: 'Choose File'
            },
            label: 'Choose File',
            placeholder: 'No file chosen',
            fileNameLabel: 'File Name',
            sizeLabel: 'Size',
            progressLabel: 'Progress'
        };
        this.sidepanel = {
            aria: {
                pinLabel: 'Pin',
                unpinLabel: 'Unpin'
            }
        };
    }
};
RufShellIntl = tslib_1.__decorate([
    Injectable()
], RufShellIntl);
export { RufShellIntl };
