// @ruf/shell/esm5/src/trigger/trigger.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Renderer2, HostListener, Input } from '@angular/core';
import { CdkOverlayOrigin } from '@angular/cdk/overlay';
import { Subscription } from 'rxjs';
import { throwTargetMissingError } from './trigger.interface';
var RufTriggerDirective = /** @class */ (function (_super) {
    tslib_1.__extends(RufTriggerDirective, _super);
    function RufTriggerDirective(elRef, renderer) {
        var _this = _super.call(this, elRef) || this;
        _this.elRef = elRef;
        _this.renderer = renderer;
        _this.subscription = Subscription.EMPTY;
        return _this;
    }
    RufTriggerDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        this._checkTarget();
        if (this.target.hasPopup()) {
            this.renderer.setAttribute(this.elRef.nativeElement, 'aria-haspopup', 'true');
            this.renderer.setAttribute(this.elRef.nativeElement, 'aria-expanded', 'false');
        }
        else {
            this.renderer.removeAttribute(this.elRef.nativeElement, 'aria-haspopup');
            this.renderer.removeAttribute(this.elRef.nativeElement, 'aria-expanded');
        }
        if (this.subscription === Subscription.EMPTY) {
            this.subscription = this.target.toggle.subscribe(function (opened) {
                _this.renderer.setAttribute(_this.elRef.nativeElement, 'aria-expanded', opened ? 'true' : 'false');
                if (!opened) {
                    _this.elRef.nativeElement.focus();
                }
            });
        }
    };
    RufTriggerDirective.prototype.handleClickEvent = function (event) {
        this.target.handleClick(event);
    };
    RufTriggerDirective.prototype.handleKeydownEvent = function (event) {
        this.target.handleKeydown(event);
    };
    RufTriggerDirective.prototype._checkTarget = function () {
        if (!this.target) {
            throwTargetMissingError();
        }
    };
    RufTriggerDirective.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    tslib_1.__decorate([
        Input('rufTrigger'),
        tslib_1.__metadata("design:type", Object)
    ], RufTriggerDirective.prototype, "target", void 0);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufTriggerDirective.prototype, "handleClickEvent", null);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufTriggerDirective.prototype, "handleKeydownEvent", null);
    RufTriggerDirective = tslib_1.__decorate([
        Directive({
            selector: '[rufTrigger]',
            exportAs: 'rufTrigger'
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], RufTriggerDirective);
    return RufTriggerDirective;
}(CdkOverlayOrigin));
export { RufTriggerDirective };
