// @ruf/material/fesm2015/ruf-material.js
/*!
FIS Responsive UI Framework: @ruf/material
Version: 6.0.0

© 2018
Fidelity National Information Services, Inc. and/or its subsidiaries - All Rights Reserved worldwide.
This document is protected under the trade secret and copyright laws as the property of Fidelity National Information Services, Inc. and/or its subsidiaries.
Copying, reproduction or distribution should be limited and only to employees with a “need to know” to do their job. Any disclosure of this document to third parties is strictly prohibited.
*/

import { __decorate, __metadata, __param } from 'tslib';
import { HostBinding, Directive, Host, NgModule } from '@angular/core';
import { MatFormField } from '@angular/material/form-field';
import { CommonModule } from '@angular/common';

// tslint:disable:directive-selector
let BoxedInputDirective = class BoxedInputDirective {
    constructor(matFormField) {
        this.matFormField = matFormField;
        this.fisStyle = '';
        matFormField.appearance = 'outline';
    }
};
__decorate([
    HostBinding('attr.fisStyle'),
    __metadata("design:type", Object)
], BoxedInputDirective.prototype, "fisStyle", void 0);
BoxedInputDirective = __decorate([
    Directive({
        selector: '[boxStyle]',
        exportAs: 'boxStyle'
    }),
    __param(0, Host()),
    __metadata("design:paramtypes", [MatFormField])
], BoxedInputDirective);

let RufBoxedInputModule = class RufBoxedInputModule {
};
RufBoxedInputModule = __decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [BoxedInputDirective],
        exports: [BoxedInputDirective]
    })
], RufBoxedInputModule);

let RufMaterialModule = class RufMaterialModule {
};
RufMaterialModule = __decorate([
    NgModule({
        imports: [RufBoxedInputModule],
        exports: [RufBoxedInputModule],
    })
], RufMaterialModule);

export { BoxedInputDirective, RufBoxedInputModule, RufMaterialModule };
