// @ruf/shell/esm5/src/menubar/menubar-item.component.js
import * as tslib_1 from "tslib";
import { RufMenubarComponent } from './menubar.component';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Input, HostBinding, HostListener, ElementRef } from '@angular/core';
import { Subject } from 'rxjs';
import { distinctUntilChanged, takeUntil } from 'rxjs/operators';
var RufMenubarItemComponent = /** @class */ (function () {
    // this can probably only be achieved in nav-menubar
    // @HostBinding('attr.aria-haspopup') hasPopup = true;
    function RufMenubarItemComponent(menubar, _elementRef) {
        this.menubar = menubar;
        this._elementRef = _elementRef;
        this.unsubscribe = new Subject();
        this.selectable = true;
        this.menuitem = true;
        this.role = 'menuitem';
        this.tabindex = -1;
        this.active = false;
    }
    RufMenubarItemComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.menubar.selectedPath$.pipe(takeUntil(this.unsubscribe), distinctUntilChanged())
            .subscribe(function (path) {
            _this.active = _this.matches(path);
            setTimeout(function () {
                _this.tabindex = _this.active ? 0 : -1;
                _this.menubar.setActiveItem(path);
            });
        });
        this.menubar.activatedPath$.pipe(takeUntil(this.unsubscribe), distinctUntilChanged())
            .subscribe(function (path) {
            var match = _this.matches(path);
            setTimeout(function () { return _this.tabindex = match ? 0 : -1; });
        });
    };
    RufMenubarItemComponent.prototype.ngOnDestroy = function () {
        this.unsubscribe.next();
        this.unsubscribe.complete();
    };
    RufMenubarItemComponent.prototype.onClick = function (event) {
        if (!event.cancelBubble) {
            if (this.selectable) {
                this.menubar.select(this.path, this.data, event);
            }
            else {
                // still change activated state since user clicked on the item
                this.menubar.activate(this.path);
            }
        }
    };
    RufMenubarItemComponent.prototype.focus = function () {
        if (this._elementRef) {
            this._elementRef.nativeElement.focus();
        }
    };
    RufMenubarItemComponent.prototype.matches = function (selectedPath) {
        if (selectedPath && this.path) {
            var pattern = new RegExp('^((\/)?' + this.path + ')(\/.*)?$'); // use word boundary to exact match
            if (selectedPath === this.path || pattern.test(selectedPath)) { // match with current menuitem and with parent menuitem
                return true;
            }
        }
        return false;
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufMenubarItemComponent.prototype, "path", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarItemComponent.prototype, "data", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarItemComponent.prototype, "selectable", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-menu-item'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarItemComponent.prototype, "menuitem", void 0);
    tslib_1.__decorate([
        HostBinding('attr.role'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarItemComponent.prototype, "role", void 0);
    tslib_1.__decorate([
        HostBinding('attr.tabindex'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarItemComponent.prototype, "tabindex", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-menubar-item-active'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarItemComponent.prototype, "active", void 0);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufMenubarItemComponent.prototype, "onClick", null);
    RufMenubarItemComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-menubar-item, [rufMenubarItem]',
            template: "\n      <ng-content></ng-content>\n  ",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        tslib_1.__metadata("design:paramtypes", [RufMenubarComponent, ElementRef])
    ], RufMenubarItemComponent);
    return RufMenubarItemComponent;
}());
export { RufMenubarItemComponent };
