// @ruf/shell/esm5/src/menubar/menubar.component.js
import * as tslib_1 from "tslib";
import { Input, Output, EventEmitter, Component, ViewEncapsulation, ChangeDetectionStrategy, ElementRef, Renderer2, HostListener, ContentChildren, QueryList, forwardRef } from '@angular/core';
import { RufMenubarItemComponent } from './menubar-item.component';
import { RufKeyboardNavigation } from '../input/keyboard';
import { BehaviorSubject, from, Subject } from 'rxjs';
import { RufMixinBase } from '../common/color/color';
var RufMenubarComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RufMenubarComponent, _super);
    function RufMenubarComponent(_elementRef, _renderer) {
        var _this = _super.call(this, _elementRef) || this;
        _this._elementRef = _elementRef;
        _this._renderer = _renderer;
        _this._selectedPathSubject = new BehaviorSubject(_this._selectedPath);
        _this._selectedPath$ = from(_this._selectedPathSubject);
        _this._activatedPathSubject = new Subject();
        _this._activatedPath$ = from(_this._activatedPathSubject);
        // if true, selecting automatically activates the item
        _this.autoActivate = true;
        _this.direction = 'row';
        /* tslint:disable:no-output-rename */
        // Renaming output to avoid duplicate `select` identifiers
        _this.selectEmitter = new EventEmitter();
        return _this;
    }
    RufMenubarComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.addNewKeyboardNav();
        setTimeout(function () { return _this.addTabIndex(); });
    };
    RufMenubarComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.menuItems.changes.subscribe(function () {
            _this.addNewKeyboardNav();
            if (_this._keyboardNav) {
                _this._keyboardNav.setActiveItemWithoutFocus(_this.selectedIndex(_this.selectedPath));
            }
        });
    };
    RufMenubarComponent.prototype.addNewKeyboardNav = function () {
        var _this = this;
        this._keyboardNav = new RufKeyboardNavigation();
        this._keyboardNav.setDirection(this.direction);
        this._keyboardNav.init(this._elementRef, this.menuItems).subscribe(function (item) { return _this.select(item.path, item.data); });
    };
    RufMenubarComponent.prototype.setActiveItem = function (path) {
        this._keyboardNav.setActiveItemWithoutFocus(this.selectedIndex(path));
        // if path is null, set tabindex to 0 for first menuitem
        if (!path) {
            this.addTabIndex();
        }
    };
    RufMenubarComponent.prototype.firstChild = function () {
        return this.menuItems ? this.menuItems.toArray()[0] : null;
    };
    RufMenubarComponent.prototype.addTabIndex = function () {
        if (this.firstChild()) {
            this.firstChild().tabindex = 0;
        }
    };
    RufMenubarComponent.prototype.selectedIndex = function (path) {
        var index = 0;
        if (path && this.menuItems) {
            index = this.menuItems.toArray().findIndex(function (menuItem) {
                return menuItem.matches(path);
            });
        }
        return index;
    };
    RufMenubarComponent.prototype.select = function (path, data, event) {
        if (!path) {
            this._selectedPath = null;
            return;
        }
        this.selectEmitter.emit({
            path: path,
            data: data,
            event: event
        });
        if (this.autoActivate) {
            this.activate(path);
        }
        this._keyboardNav.focus(path);
        this._activatedPathSubject.next(path);
    };
    RufMenubarComponent.prototype.activate = function (path) {
        this._selectedPath = path;
        this._selectedPathSubject.next(path);
    };
    Object.defineProperty(RufMenubarComponent.prototype, "selectedPath", {
        get: function () {
            return this._selectedPath;
        },
        // Selection of an item is driven by this input if it is used.
        // Otherwise, the select method (which is invoked on click of an item) changes the selectedPath automatically.
        // If this property setter is invoked, it turns off automatic selection.
        // Normally, this should only happen when a selectedPath property binding is used.
        set: function (path) {
            this.autoActivate = false;
            this.activate(path);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufMenubarComponent.prototype, "selectedPath$", {
        get: function () {
            return this._selectedPath$;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufMenubarComponent.prototype, "activatedPath$", {
        get: function () {
            return this._activatedPath$;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufMenubarComponent.prototype, "inkbarPosition", {
        get: function () {
            return this._inkbarPosition;
        },
        set: function (position) {
            this._updateInkbarPosition(position === 'start' ? position : 'end');
        },
        enumerable: true,
        configurable: true
    });
    RufMenubarComponent.prototype.getLayoutClass = function () {
        return {
            'ruf-menubar-layout': true,
            'ruf-layout-row': this.direction !== 'column',
            'ruf-layout-column': this.direction === 'column',
            'ruf-layout-row-xs': this.directionXs === 'row',
            'ruf-layout-column-xs': this.directionXs === 'column',
            'ruf-layout-row-sm': this.directionSm === 'row',
            'ruf-layout-column-sm': this.directionSm === 'column',
            'ruf-layout-row-md': this.directionMd === 'row',
            'ruf-layout-column-md': this.directionMd === 'column',
            'ruf-layout-row-lg': this.directionLg === 'row',
            'ruf-layout-column-lg': this.directionLg === 'column',
            'ruf-layout-row-xl': this.directionXl === 'row',
            'ruf-layout-column-xl': this.directionXl === 'column',
            'ruf-layout-row-gt-xs': this.directionGtXs === 'row',
            'ruf-layout-column-gt-xs': this.directionGtXs === 'column',
            'ruf-layout-row-gt-sm': this.directionGtSm === 'row',
            'ruf-layout-column-gt-sm': this.directionGtSm === 'column',
            'ruf-layout-row-gt-md': this.directionGtMd === 'row',
            'ruf-layout-column-gt-md': this.directionGtMd === 'column',
            'ruf-layout-row-gt-lg': this.directionGtLg === 'row',
            'ruf-layout-column-gt-lg': this.directionGtLg === 'column',
            'ruf-layout-row-lt-sm': this.directionLtSm === 'row',
            'ruf-layout-column-lt-sm': this.directionLtSm === 'column',
            'ruf-layout-row-lt-md': this.directionLtMd === 'row',
            'ruf-layout-column-lt-md': this.directionLtMd === 'column',
            'ruf-layout-row-lt-lg': this.directionLtLg === 'row',
            'ruf-layout-column-lt-lg': this.directionLtLg === 'column',
            'ruf-layout-row-lt-xl': this.directionLtXl === 'row',
            'ruf-layout-column-lt-xl': this.directionLtXl === 'column'
        };
    };
    RufMenubarComponent.prototype._updateInkbarPosition = function (position) {
        this._renderer.removeClass(this._elementRef.nativeElement, "ruf-menubar-inkbar-" + this._inkbarPosition);
        this._renderer.addClass(this._elementRef.nativeElement, "ruf-menubar-inkbar-" + position);
        this._inkbarPosition = position;
    };
    RufMenubarComponent.prototype.onKey = function (event) {
        if (this._keyboardNav && this._keyboardNav.isActive()) {
            this._keyboardNav.onKeyDown(event);
        }
    };
    tslib_1.__decorate([
        Input('activeColor'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "color", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "direction", void 0);
    tslib_1.__decorate([
        Input('direction.xs'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionXs", void 0);
    tslib_1.__decorate([
        Input('direction.sm'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionSm", void 0);
    tslib_1.__decorate([
        Input('direction.md'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionMd", void 0);
    tslib_1.__decorate([
        Input('direction.lg'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionLg", void 0);
    tslib_1.__decorate([
        Input('direction.xl'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionXl", void 0);
    tslib_1.__decorate([
        Input('direction.gt-xs'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionGtXs", void 0);
    tslib_1.__decorate([
        Input('direction.gt-sm'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionGtSm", void 0);
    tslib_1.__decorate([
        Input('direction.gt-md'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionGtMd", void 0);
    tslib_1.__decorate([
        Input('direction.gt-lg'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionGtLg", void 0);
    tslib_1.__decorate([
        Input('direction.lt-sm'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionLtSm", void 0);
    tslib_1.__decorate([
        Input('direction.lt-md'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionLtMd", void 0);
    tslib_1.__decorate([
        Input('direction.lt-lg'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionLtLg", void 0);
    tslib_1.__decorate([
        Input('direction.lt-xl'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "directionLtXl", void 0);
    tslib_1.__decorate([
        Output('select'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenubarComponent.prototype, "selectEmitter", void 0);
    tslib_1.__decorate([
        ContentChildren(forwardRef(function () { return RufMenubarItemComponent; })),
        tslib_1.__metadata("design:type", QueryList)
    ], RufMenubarComponent.prototype, "menuItems", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufMenubarComponent.prototype, "selectedPath", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufMenubarComponent.prototype, "inkbarPosition", null);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [KeyboardEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufMenubarComponent.prototype, "onKey", null);
    RufMenubarComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-menubar',
            template: "<div [ngClass]=\"getLayoutClass()\" role=\"menubar\">\n  <ng-content></ng-content>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["ruf-menubar{-ms-grid-row-align:stretch;align-self:stretch;display:block}.ruf-menubar-layout{-webkit-box-align:stretch;align-items:stretch;height:100%}.ruf-menu-item{-webkit-box-align:center;align-items:center;border-color:transparent;cursor:pointer;display:-webkit-box;display:flex;-webkit-box-flex:1;flex:1 auto;-webkit-transition:background-color .2s;transition:background-color .2s}.ruf-menu-item:focus{outline:0}.ruf-layout-row{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-menubar-layout.ruf-layout-column{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-menubar-layout.ruf-layout-column>ul{padding-left:0}.ruf-menubar-layout.ruf-layout-column .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}@media (min-width:0) and (max-width:599px){.ruf-layout-row-xs{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-xs>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-xs .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-xs{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-xs>ul{padding-left:0}.ruf-layout-column-xs .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:600px) and (max-width:959px){.ruf-layout-row-sm{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-sm>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-sm .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-sm{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-sm>ul{padding-left:0}.ruf-layout-column-sm .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:960px) and (max-width:1279px){.ruf-layout-row-md{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-md>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-md .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-md{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-md>ul{padding-left:0}.ruf-layout-column-md .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:1280px) and (max-width:1919px){.ruf-layout-row-lg{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-lg>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-lg .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-lg{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-lg>ul{padding-left:0}.ruf-layout-column-lg .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:1920px){.ruf-layout-row-xl{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-xl>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-xl .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-xl{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-xl>ul{padding-left:0}.ruf-layout-column-xl .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:600px){.ruf-layout-row-gt-xs{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-gt-xs>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-gt-xs .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-gt-xs{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-gt-xs>ul{padding-left:0}.ruf-layout-column-gt-xs .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:960px){.ruf-layout-row-gt-sm{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-gt-sm>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-gt-sm .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-gt-sm{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-gt-sm>ul{padding-left:0}.ruf-layout-column-gt-sm .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:1280px){.ruf-layout-row-gt-md{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-gt-md>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-gt-md .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-gt-md{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-gt-md>ul{padding-left:0}.ruf-layout-column-gt-md .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}@media (min-width:1920px){.ruf-layout-row-gt-lg{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.ruf-layout-row-gt-lg>ul{padding-left:0;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;flex-wrap:wrap;width:100%}.ruf-layout-row-gt-lg .ruf-menu-item{border-bottom-width:.125rem;border-top-width:.125rem;-webkit-box-pack:center;justify-content:center;border-style:solid none;padding:.125rem .9375rem}.ruf-layout-column-gt-lg{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.ruf-layout-column-gt-lg>ul{padding-left:0}.ruf-layout-column-gt-lg .ruf-menu-item{border-left-width:.125rem;border-right-width:.125rem;-webkit-box-pack:start;justify-content:flex-start;border-style:none solid;padding:.5rem .9375rem}}"]
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], RufMenubarComponent);
    return RufMenubarComponent;
}(RufMixinBase));
export { RufMenubarComponent };
