// @ruf/bundler/lib/cli.js
#!/usr/bin/env node
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require("./");
const util_1 = require("./util");
const path_1 = require("path");
const static_server_1 = require("./static-server");
const meow = require("meow");
const TSCONFIG = 'tsconfig.json';
function getCli() {
    return meow(`
	Usage
	  $ ruf-bundler [examplePath]

	Options
    --no-progress, -n  Do not show progress
    --open, -o         Start a server and open browser
    --serve, -s        Start a server
    --moduleCache, -mc Path to the node_modules cache
    --cache, -c        Use caching of projects
    --destination, -d  Project build destination
    --version          Prints out the version of this tool
    --help             Usage
    --dev              see Dev Mode
    --assets           Path to asset files
    --packageManager   Package Manager (pnpm or npm) used to install dependencies. Defaults to pnpm.
    --                 Pass additional flags to the template's build after --

  Dev Mode
    This will allow you to pass typescript config to the template so that your local library can be used while building.

  Examples
    $ ruf-bundler example/splitter
      # runs in the example/splitter folder if it contains an example.json file.
    $ ruf-bundler -c --open
      # runs in current folder, serves the resulting app (from cache if already built)
      # and opens the browser to see it
    $ ruf-bundler example/show -c -o -- -bh /run/\${hash}
      # passes flags to the template's build script. Here \${hash} will be replaced by the actual 'hash' code generated using example object.

`, {
        flags: {
            'no-progress': { type: 'boolean', alias: 'n' },
            open: { type: 'boolean', alias: 'o' },
            serve: { type: 'boolean', alias: 's' },
            moduleCache: { type: 'string', alias: 'mc' },
            cache: { type: 'boolean', alias: 'c' },
            destination: { type: 'string', alias: 'd' },
            workspace: { type: 'string', alias: 'w' },
            keepWorkspace: { type: 'boolean', alias: 'k' },
            dev: { type: 'boolean' },
            assets: { type: 'string' }
        }
    });
}
function getTemplateChanges(workdir, flags) {
    return __awaiter(this, void 0, void 0, function* () {
        if (flags.dev) {
            const localDir = yield util_1.findup(workdir, TSCONFIG).catch(_ => '');
            if (localDir) {
                const tsconf = require(path_1.join(localDir, TSCONFIG));
                const paths = tsconf.compilerOptions && tsconf.compilerOptions.paths;
                if (paths) {
                    Object.keys(paths).forEach(p => {
                        paths[p] = paths[p].map((dest) => path_1.resolve(localDir, dest));
                    });
                    return {
                        merge: {
                            'tsconfig.json': {
                                compilerOptions: {
                                    paths: paths
                                }
                            },
                            'package.json': {
                                scripts: {
                                    build: 'ng build'
                                }
                            }
                        }
                    };
                }
            }
        }
        return undefined;
    });
}
function main() {
    return __awaiter(this, void 0, void 0, function* () {
        const cli = getCli();
        const workdir = (cli.input[0] && path_1.resolve(process.cwd(), cli.input[0])) || process.cwd();
        const exampleFile = path_1.join(workdir, 'ruf-example.json');
        const isExampleDir = yield util_1.exists(exampleFile);
        if (!isExampleDir) {
            throw new Error(`${workdir} does not contain an example.`);
        }
        const example = require(exampleFile);
        const bundler = new _1.Bundler({ cachePath: cli.flags.moduleCache });
        console.log(`Modules cache: ${bundler.cachePath}`);
        const template = bundler.getTemplate(example.template);
        const loadedFiles = yield util_1.getLoadedFiles({
            cwd: workdir,
            assets: cli.flags.assets
        }, example.ignore);
        const templateChanges = yield getTemplateChanges(workdir, cli.flags);
        try {
            const path = yield template.build(Object.assign({}, example, { files: Object.assign({}, example.files, loadedFiles) }), {
                progress: cli.flags['no-progress']
                    ? util_1.noop
                    : ([step, total, msg, hash, done]) => console.log(`${step}/${total} ==> ${msg}`),
                buildPlan: cli.flags['no-progress']
                    ? util_1.noop
                    : plan => Object.keys(plan).forEach(p => console.log(`${p}: ${plan[p]}`)),
                useHashDir: cli.flags.cache,
                destination: cli.flags.destination &&
                    path_1.resolve(process.cwd(), cli.flags.destination),
                keepWorkspace: cli.flags.keepWorkspace,
                workspace: cli.flags.workspace && path_1.resolve(process.cwd(), cli.flags.workspace),
                returnIfExists: cli.flags.cache,
                templateChanges,
                buildFlags: cli.input.slice(1).join(' '),
                packageManager: cli.flags.packageManager
            });
            console.log(`\nApplication created at ${path}`);
            if (cli.flags.serve || cli.flags.open) {
                console.log(`Serving at http://localhost:8080`);
                console.log(`CTRL-C to quit.`);
                static_server_1.serve(path, cli.flags.open);
            }
        }
        catch (err) {
            console.log(err.message);
            if (err.details) {
                console.log(err.details);
            }
        }
    });
}
main();
