// @ruf/bundler/lib/dependency-manager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const execa = require('execa');
const util_1 = require("./util");
const os_1 = require("os");
const isWin = os_1.platform() === 'win32';
const { getHash, noop, getDirs, move, copy, rm, mkdir, writeFile, symlink, exists } = require('./util');
class DependencyManager {
    constructor(opts = {}) {
        this.counter = 0;
        this._cachePath = opts.cachePath || process.env.RUF_BUNDLER_MODULES_PATH;
        if (this._cachePath === null) {
            throw new Error('A cachePath option needs to be defined. Use RUF_BUNDLER_MODULES_PATH env var if necessary.');
        }
    }
    get cachePath() {
        return this._cachePath;
    }
    createNodeModules(hash, path, templatePath, deps) {
        return __awaiter(this, void 0, void 0, function* () {
            yield mkdir(path);
            yield writeFile(path_1.join(path, 'package.json'), JSON.stringify({
                name: hash,
                description: templatePath,
                repository: 'none',
                version: '0.0.0',
                dependencies: deps
            }));
            const toCopy = Object.values(deps)
                .filter(v => v.indexOf('file:') === 0)
                .map(f => f.substring(5));
            for (const f of toCopy) {
                yield copy(path_1.join(templatePath, f), path_1.join(path, f));
            }
            yield execa('npm', ['install', '--no-progress'], {
                cwd: path
            }).catch((err) => __awaiter(this, void 0, void 0, function* () {
                yield rm(path);
                throw new util_1.DetailedError('Failed installing dependencies', err.message, hash);
            }));
            return path;
        });
    }
    linkDeps(src, dest, scoped) {
        return __awaiter(this, void 0, void 0, function* () {
            const dirs = yield getDirs(src);
            for (const dir of dirs) {
                const target = isWin ? path_1.resolve(path_1.join(src, dir)) : path_1.join(src, dir);
                const ln = isWin ? path_1.resolve(path_1.join(dest, dir)) : path_1.join(dest, dir);
                try {
                    if (!scoped && dir.charAt(0) !== '@') {
                        yield symlink(target, ln, 'junction');
                    }
                    else {
                        yield copy(target, ln);
                    }
                }
                catch (err) {
                }
                if (!scoped && dir.charAt(0) === '@') {
                    yield this.linkDeps(target, ln, true);
                }
            }
        });
    }
    maybeInstallDeps(path, templatePath, deps, opts = { link: false }) {
        return __awaiter(this, void 0, void 0, function* () {
            const hash = getHash(deps);
            const progress = opts.progress || noop;
            const installPath = path_1.join(path, hash);
            const tempPath = isWin
                ? installPath
                : `${installPath}-${new Date().getTime()}-${this.counter++}.temp`;
            const nodeModulesPath = path_1.join(installPath, 'node_modules');
            const ex = yield exists(nodeModulesPath);
            if (!ex) {
                progress('Installing dependencies. This may take a while.', hash);
                yield this.createNodeModules(hash, tempPath, templatePath, deps);
                if (opts.link) {
                    yield this.linkDeps(path_1.join(path, 'node_modules'), path_1.join(tempPath, 'node_modules'));
                }
                try {
                    yield move(tempPath, installPath);
                }
                catch (err) {
                    console.error('Error moving temp modules path to final location:', err);
                    yield rm(tempPath);
                }
            }
            else {
                progress('Dependencies already installed.', hash);
            }
            return installPath;
        });
    }
    getNodeModulesPathForTemplate(template, deps = {}, opts = {}, hash) {
        return __awaiter(this, void 0, void 0, function* () {
            const progress = opts.progress || noop;
            progress('Preparing dev dependencies.', hash);
            const depDepsPath = yield this.maybeInstallDeps(this._cachePath, template.templateSrc, template.devDependencies, { link: false, progress: progress });
            progress('Preparing project dependencies.', hash);
            const depsPath = yield this.maybeInstallDeps(depDepsPath, template.templateSrc, Object.assign({}, template.dependencies, deps), { link: true, progress: progress });
            return path_1.join(depsPath, 'node_modules');
        });
    }
}
exports.DependencyManager = DependencyManager;
