// @ruf/bundler/lib/util.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const crypto = require("crypto");
const fs = require("fs-extra");
const nodefs = require("fs");
const globOrig = require("glob");
const rimraf = require("rimraf");
const matches = require("minimatch");
const mkdirp = require("mkdirp");
const path_1 = require("path");
const dm = require("deepmerge");
const execa = require("execa");
var os_1 = require("os");
exports.tmpdir = os_1.tmpdir;
exports.lstat = util_1.promisify(fs.lstat);
exports.readdir = util_1.promisify(fs.readdir);
exports.copy = fs.copy;
exports.move = fs.move;
exports.readFile = util_1.promisify(fs.readFile);
exports.writeFile = util_1.promisify(fs.writeFile);
exports.rename = util_1.promisify(fs.rename);
exports.rm = util_1.promisify(rimraf);
exports.mkdir = util_1.promisify(mkdirp);
exports.symlink = util_1.promisify(nodefs.symlink);
exports.access = util_1.promisify(fs.access);
exports.glob = util_1.promisify(globOrig);
exports.findup = util_1.promisify(require('findup'));
class DetailedError extends Error {
    constructor(msg, details, hash) {
        super(msg);
        this.details = details;
        this.hash = hash;
    }
}
exports.DetailedError = DetailedError;
function readJSONFile(file) {
    return __awaiter(this, void 0, void 0, function* () {
        return JSON.parse((yield exports.readFile(file)).toString());
    });
}
exports.readJSONFile = readJSONFile;
function writeJSONFile(file, contents) {
    return __awaiter(this, void 0, void 0, function* () {
        return exports.writeFile(file, JSON.stringify(contents, null, 2));
    });
}
exports.writeJSONFile = writeJSONFile;
function getSHA1(data) {
    return crypto
        .createHash('sha1')
        .update(data)
        .digest('hex');
}
exports.getSHA1 = getSHA1;
function getHash(obj) {
    if (Array.isArray(obj)) {
        return getSHA1(obj.sort().join(','));
    }
    if (typeof obj === 'object') {
        return getSHA1(Object.keys(obj)
            .sort()
            .map(k => `${k}:${getHash(obj[k])}`)
            .join(';'));
    }
    return obj.toString();
}
exports.getHash = getHash;
function noop() { }
exports.noop = noop;
function isDirectory(path, source) {
    return __awaiter(this, void 0, void 0, function* () {
        return (yield exports.lstat(path_1.join(path, source))).isDirectory();
    });
}
exports.isDirectory = isDirectory;
function getDirs(path) {
    return __awaiter(this, void 0, void 0, function* () {
        return (yield exports.readdir(path)).filter(isDirectory.bind(null, path));
    });
}
exports.getDirs = getDirs;
function exists(path) {
    return __awaiter(this, void 0, void 0, function* () {
        return exports.access(path)
            .then(_ => true)
            .catch(e => false);
    });
}
exports.exists = exists;
function writeFileForce(path, contents) {
    return __awaiter(this, void 0, void 0, function* () {
        return exports.mkdir(path_1.dirname(path)).then(_ => exports.writeFile(path, contents));
    });
}
exports.writeFileForce = writeFileForce;
function isObject(a) {
    return !isArray(a) && typeof a === 'object' && a !== null;
}
exports.isObject = isObject;
function isArray(a) {
    return Array.isArray(a);
}
exports.isArray = isArray;
function isString(a) {
    return typeof a === 'string';
}
exports.isString = isString;
exports.deepmerge = dm;
function changeJSONFiles(dir, mergers) {
    return __awaiter(this, void 0, void 0, function* () {
        if (isObject(mergers)) {
            const files = Object.keys(mergers);
            for (const file of files) {
                const absfile = path_1.join(dir, file);
                const contents = yield readJSONFile(absfile);
                const newContents = exports.deepmerge(contents, mergers[file]);
                yield writeJSONFile(absfile, newContents);
            }
        }
    });
}
exports.changeJSONFiles = changeJSONFiles;
function linkDependencies(path, linkDeps, hash) {
    return __awaiter(this, void 0, void 0, function* () {
        if (linkDeps && linkDeps.length > 0) {
            for (const dep of linkDeps) {
                yield execa('npm', ['link', dep], {
                    cwd: path
                }).catch((err) => {
                    throw new DetailedError(`Linking ${dep} failed`, err.message, hash);
                });
            }
        }
    });
}
exports.linkDependencies = linkDependencies;
function getLoadedFiles(opts, ignoreFiles) {
    return __awaiter(this, void 0, void 0, function* () {
        let files = (yield exports.glob('**/*.@(ts|scss|css|html|jpg|png|md|js|jsx|tsx|json)', { cwd: opts.cwd })).filter(f => !f.endsWith('ruf-example.json'));
        if (ignoreFiles && ignoreFiles.length > 0) {
            files = files.filter(f => !ignoreFiles.some((ignore) => matches(f, ignore)));
        }
        const loadedFiles = {};
        for (const f of files) {
            loadedFiles[f] = (yield exports.readFile(path_1.join(opts.cwd, f))).toString();
        }
        if (opts.assets) {
            const assetPath = path_1.join(opts.cwd, opts.assets);
            const assetDir = assetPath.split(path_1.sep).pop();
            files = yield exports.glob('**/*.*', { cwd: assetPath });
            for (const f of files) {
                loadedFiles[`assets/${assetDir}/${f}`] = (yield exports.readFile(path_1.join(assetPath, f))).toString();
            }
        }
        return loadedFiles;
    });
}
exports.getLoadedFiles = getLoadedFiles;
