// @ruf/shell/esm5/src/footer/footer.component.js
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation, Directive, ContentChild, HostBinding } from '@angular/core';
var RufFooterControlsDirective = /** @class */ (function () {
    function RufFooterControlsDirective() {
    }
    RufFooterControlsDirective = tslib_1.__decorate([
        Directive({
            selector: '[rufFooterControls]'
        })
    ], RufFooterControlsDirective);
    return RufFooterControlsDirective;
}());
export { RufFooterControlsDirective };
var RufFooterComponent = /** @class */ (function () {
    function RufFooterComponent() {
        this.role = 'contentinfo';
    }
    RufFooterComponent.prototype.ngOnInit = function () { };
    tslib_1.__decorate([
        HostBinding('attr.role'),
        tslib_1.__metadata("design:type", Object)
    ], RufFooterComponent.prototype, "role", void 0);
    tslib_1.__decorate([
        ContentChild(RufFooterControlsDirective, { static: false }),
        tslib_1.__metadata("design:type", RufFooterControlsDirective)
    ], RufFooterComponent.prototype, "controls", void 0);
    RufFooterComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-footer',
            template: "<ruf-toolbar class=\"ruf-footer-toolbar\" rufId=\"footer_toolbar\">\n  <ng-content></ng-content>\n  <div rufToolbarFiller *ngIf=\"controls\" class=\"ruf-footer-filler\"></div>\n  <ng-content select=\"[rufFooterControls]\"></ng-content>\n</ruf-toolbar>\n",
            encapsulation: ViewEncapsulation.None,
            styles: ["ruf-footer{border-top-style:solid;border-top-width:1px;display:block;flex-shrink:0;min-height:2.5rem;overflow:hidden}.ruf-footer-toolbar{height:2.5rem;margin:0 auto;padding:0 .625rem;width:100%}.ruf-footer-filler{-webkit-box-flex:1;flex:1 1 auto}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], RufFooterComponent);
    return RufFooterComponent;
}());
export { RufFooterComponent };
