// @ruf/shell/esm2015/src/ellipsis/ellipsis.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild, ElementRef, Input, Output, EventEmitter, ViewEncapsulation, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { RufShellIntl } from '../utils/shell-intl';
import { BehaviorSubject } from 'rxjs';
let RufEllipsisComponent = class RufEllipsisComponent {
    constructor(_intl, changeDetectorRef) {
        this._intl = _intl;
        this.changeDetectorRef = changeDetectorRef;
        this.showReadMore$ = new BehaviorSubject(false);
        this.lines = 5;
        this.lineHeight = 1;
        this.expand = new EventEmitter();
        this._intlChanges = _intl.changes.subscribe(() => this.changeDetectorRef.markForCheck());
    }
    get label() {
        return this._label || this._intl.ellipsis.label || '...';
    }
    set label(value) {
        this._label = value;
    }
    get ariaLabel() {
        return this._ariaLabel || this._intl.ellipsis.aria.label;
    }
    set ariaLabel(value) {
        this._ariaLabel = value;
    }
    checkReadMore() {
        if (this.text && this.ellipsisContainer.nativeElement.offsetHeight < this.ellipsisContainer.nativeElement.scrollHeight) {
            this.showReadMore$.next(true);
        }
        else {
            this.showReadMore$.next(false);
        }
    }
    ngAfterViewChecked() {
        this.checkReadMore();
        this.changeDetectorRef.detectChanges();
    }
    ngOnDestroy() {
        this._intlChanges.unsubscribe();
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], RufEllipsisComponent.prototype, "text", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufEllipsisComponent.prototype, "lines", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufEllipsisComponent.prototype, "lineHeight", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufEllipsisComponent.prototype, "label", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufEllipsisComponent.prototype, "ariaLabel", null);
tslib_1.__decorate([
    ViewChild('ellipsisContainer', { static: false }),
    tslib_1.__metadata("design:type", ElementRef)
], RufEllipsisComponent.prototype, "ellipsisContainer", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], RufEllipsisComponent.prototype, "expand", void 0);
RufEllipsisComponent = tslib_1.__decorate([
    Component({
        selector: 'ruf-ellipsis',
        template: "<div  #ellipsisContainer [class]=\"'ellipsis-container'\" [style.maxHeight.rem]=\"lines * lineHeight\" [style.lineHeight.rem]=\"lineHeight\">\n  <div  class=\"ellipsis-wrapper\" >\n    <div class=\"ellipsis-content\" [innerHTML]=\"text\">\n    </div>\n  </div>\n  <span *ngIf=\"showReadMore$ | async\" [attr.aria-label]=\"ariaLabel\" class=\"ellipsis\">\n    <a (click)=\"expand.emit({'text': this.text});\" href=\"javascript:;\">{{label}}</a>\n  </span>\n</div>\n",
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: ["ruf-ellipsis .ellipsis-container{max-height:calc(5rem);overflow:hidden;position:relative}ruf-ellipsis .ellipsis-container:before{content:'';float:left;height:100%;width:1px}ruf-ellipsis .ellipsis-wrapper{float:right;margin-left:-1px;width:100%}ruf-ellipsis .ellipsis-content{margin:0}ruf-ellipsis .ellipsis-content p{margin-bottom:1rem;margin-top:0}ruf-ellipsis .ellipsis{bottom:0;position:absolute;right:0;text-align:right;width:auto}"]
    }),
    tslib_1.__metadata("design:paramtypes", [RufShellIntl, ChangeDetectorRef])
], RufEllipsisComponent);
export { RufEllipsisComponent };
