// @ruf/shell/esm5/src/card/card.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, ChangeDetectionStrategy, Component, Directive, ViewEncapsulation, Input } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { RufShellIntl } from '../utils/shell-intl';
var RufCardControlsDirective = /** @class */ (function () {
    function RufCardControlsDirective() {
    }
    RufCardControlsDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-card-controls' // tslint:disable-line
        })
    ], RufCardControlsDirective);
    return RufCardControlsDirective;
}());
export { RufCardControlsDirective };
var RufCardSubtitleDirective = /** @class */ (function () {
    function RufCardSubtitleDirective() {
    }
    RufCardSubtitleDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-card-subtitle' // tslint:disable-line
        })
    ], RufCardSubtitleDirective);
    return RufCardSubtitleDirective;
}());
export { RufCardSubtitleDirective };
var RufCardComponent = /** @class */ (function () {
    function RufCardComponent(_intl, changeDetectorRef) {
        this._intl = _intl;
        this.expandable = false;
        this.expanded = true;
        this._intlChanges = _intl.changes.subscribe(function () { return changeDetectorRef.markForCheck(); });
    }
    Object.defineProperty(RufCardComponent.prototype, "expandLabel", {
        get: function () {
            return this._expandLabel || this._intl.card.aria.expandLabel;
        },
        set: function (value) {
            this._expandLabel = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufCardComponent.prototype, "collapseLabel", {
        get: function () {
            return this._collapseLabel || this._intl.card.aria.collapseLabel;
        },
        set: function (value) {
            this._collapseLabel = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufCardComponent.prototype, "toggleLabel", {
        get: function () {
            return this.expanded ? this.collapseLabel : this.expandLabel;
        },
        enumerable: true,
        configurable: true
    });
    RufCardComponent.prototype.ngOnDestroy = function () {
        this._intlChanges.unsubscribe();
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufCardComponent.prototype, "title", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufCardComponent.prototype, "expandable", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufCardComponent.prototype, "expanded", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufCardComponent.prototype, "expandLabel", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufCardComponent.prototype, "collapseLabel", null);
    RufCardComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-card',
            template: "<mat-card class=\"ruf-card\" rufId=\"card_panel\" [ngClass]=\"{'ruf-card-collapsed': !expanded, 'ruf-card-expanded': expanded}\">\n  <mat-card-header class=\"ruf-card-header\" rufId=\"card_header\">\n    <button type=\"button\" mat-icon-button *ngIf=\"expandable\" (click)=\"expanded = !expanded\" class=\"ruf-card-expand\" [attr.aria-label]=\"toggleLabel\">\n      <mat-icon fisIcon=\"chevron\"></mat-icon>\n    </button>\n    <mat-card-title-group class=\"ruf-card-titlegroup\">\n      <mat-card-title class=\"ruf-card-title\" (click)=\"expandable && expanded = !expanded\">{{title}}</mat-card-title>\n      <mat-card-subtitle class=\"ruf-card-subtitle\">\n        <ng-content select=\"ruf-card-subtitle\"></ng-content>\n      </mat-card-subtitle>\n    </mat-card-title-group>\n    <div class=\"ruf-card-filler\"></div>\n    <div class=\"ruf-card-controls\">\n      <ng-content select=\"ruf-card-controls\"></ng-content>\n    </div>\n  </mat-card-header>\n  <mat-card-content class=\"ruf-card-content\" rufId=\"card_content\" [@expanded]=\"'' + expanded\">\n    <ng-content></ng-content>\n    <ng-content select=\"ruf-card-action-bar\"></ng-content>\n  </mat-card-content>\n</mat-card>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [
                trigger('expanded', [
                    state('false', style({
                        height: 0,
                        visibility: 'hidden'
                    })),
                    state('true', style({
                        height: '*',
                        visibility: 'visible'
                    })),
                    transition('false => true', animate('300ms ease-in-out')),
                    transition('true => false', animate('300ms ease-in-out'))
                ])
            ],
            styles: ["ruf-card{display:-webkit-box;display:flex}.ruf-card.mat-card{border-radius:2px;margin:0;padding:0}.ruf-card.mat-card .ruf-card-header{min-height:2.5rem;padding-left:12px}.ruf-card.mat-card .ruf-card-expand{cursor:pointer;height:30px;line-height:30px;margin:7px 5px 5px -5px;-webkit-transform:rotate(180deg);transform:rotate(180deg);-webkit-transition:-webkit-transform .3s ease-out;transition:transform .3s ease-out;transition:transform .3s ease-out,-webkit-transform .3s ease-out;width:30px}.ruf-card.mat-card.ruf-card-collapsed .ruf-card-expand{-webkit-transform:rotate(0);transform:rotate(0)}.ruf-card.mat-card .mat-select-underline{display:none}.ruf-card.mat-card .ruf-card-content{overflow:hidden}.ruf-card.mat-card .mat-card-header-text{margin:0}.ruf-card.mat-card .ruf-card-titlegroup{-ms-grid-row-align:center;align-self:center;margin:0;padding:5px 5px 5px 0}.ruf-card.mat-card .ruf-card-subtitle,.ruf-card.mat-card .ruf-card-title{margin-bottom:0}.ruf-card.mat-card .ruf-card-metric,.ruf-card.mat-card .ruf-card-metric-description{padding-right:10px;text-align:right}.ruf-card.mat-card .ruf-card-subtitle:not(:first-child),.ruf-card.mat-card .ruf-card-title:not(:first-child){margin-top:0}.ruf-card.mat-card .ruf-card-filler{-webkit-box-flex:1;flex:1 1 auto}.ruf-card.mat-card .ruf-card-controls{-ms-grid-row-align:center;align-self:center;-webkit-box-flex:0;flex:0 0 auto;margin-left:20px;padding:5px}"]
        }),
        tslib_1.__metadata("design:paramtypes", [RufShellIntl, ChangeDetectorRef])
    ], RufCardComponent);
    return RufCardComponent;
}());
export { RufCardComponent };
