// @ruf/shell/esm5/src/menu-label/menu-label.component.js
import * as tslib_1 from "tslib";
import { Input, Output, EventEmitter, Component, ViewEncapsulation, ChangeDetectionStrategy, HostBinding, HostListener, ViewChild } from '@angular/core';
import { ENTER, SPACE, ESCAPE } from '@angular/cdk/keycodes';
import { MatMenuTrigger } from '@angular/material/menu';
var RufMenuLabelComponent = /** @class */ (function () {
    function RufMenuLabelComponent() {
        this.label = '';
        this.icon = '';
        // on child menu toggle
        this.toggle = new EventEmitter();
        this.select = new EventEmitter();
        this.rowLayout = true; // tells the internal labeled-icon that it is in a row.
    }
    RufMenuLabelComponent.prototype.handleClick = function (event) {
        // labels with children will not propagate the click event by default
        if (this.children && this.children.length > 0) {
            event.preventDefault();
            event.stopPropagation();
        }
    };
    RufMenuLabelComponent.prototype.hasPopup = function () {
        return this.children ? this.children.length > 0 : false;
    };
    RufMenuLabelComponent.prototype.onClose = function () {
        this.toggle.emit(false);
    };
    RufMenuLabelComponent.prototype.handleKeydown = function (event) {
        var _this = this;
        if (!this.children || this.children.length === 0) {
            return;
        }
        switch (event.keyCode) {
            case SPACE:
            case ENTER:
                this.trigger.openMenu();
                if (this.trigger.menu) {
                    // This is needed for dynamic menubar since it uses setTimeout for selecting a tab.
                    // This should be removed when we refactor dynamic menubar.
                    setTimeout(function () {
                        _this.trigger.menu.focusFirstItem();
                    });
                }
                event.preventDefault();
                event.stopPropagation();
                this.toggle.emit(true);
                break;
            case ESCAPE:
                this.trigger.closeMenu();
                this.toggle.emit(false);
                event.stopPropagation();
                break;
            default:
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMenuLabelComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufMenuLabelComponent.prototype, "icon", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Array)
    ], RufMenuLabelComponent.prototype, "children", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], RufMenuLabelComponent.prototype, "toggle", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], RufMenuLabelComponent.prototype, "select", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-layout-row'),
        tslib_1.__metadata("design:type", Object)
    ], RufMenuLabelComponent.prototype, "rowLayout", void 0);
    tslib_1.__decorate([
        ViewChild(MatMenuTrigger, { static: false }),
        tslib_1.__metadata("design:type", MatMenuTrigger)
    ], RufMenuLabelComponent.prototype, "trigger", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], RufMenuLabelComponent.prototype, "selectedPath", void 0);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufMenuLabelComponent.prototype, "handleClick", null);
    RufMenuLabelComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-menu-label',
            template: "<div class=\"ruf-menu-label-wrap\" rufId>\n  <!-- ideally, Angular would support putting matMenuTriggerFor on the host -->\n  <div class=\"ruf-menu-label-trigger\" *ngIf=\"children && children.length > 0\" [matMenuTriggerFor]=\"menu\" (menuClosed)=\"toggle.emit(false)\" (menuOpened)=\"toggle.emit(true)\"></div>\n  <ruf-labeled-icon fontSet=\"fisfont\" [fontIcon]=\"icon\" rufId>\n    <span class=\"ruf-menu-label-container\">\n      {{ label }}\n      <mat-icon class=\"ruf-menu-label-chevron\" aria-hidden=\"true\" rufId=\"menu_label_link\" *ngIf=\"children && children.length > 0\" fisIcon=\"arrow-down\"></mat-icon>\n    </span>\n\n    <span (click)=\"$event.stopPropagation()\" rufId=\"menu_label_content\">\n      <ng-content></ng-content>\n    </span>\n  </ruf-labeled-icon>\n</div>\n\n<mat-menu class=\"ruf-cust-menu\" rufId=\"menu\" #menu=\"matMenu\" yPosition=\"below\" [overlapTrigger]=\"false\">\n  <button type=\"button\" rufId=\"menu_button\" *ngFor=\"let child of children\" mat-menu-item (click)=\"select.emit(child)\"\n          [class.ruf-menu-selected]=\"selectedPath && child.path === selectedPath\"\n          [attr.aria-selected]=\"selectedPath && child.path === selectedPath\">\n    <mat-icon rufId=\"menu_icon\" *ngIf=\"child.icon\" fontSet=\"fisfont\" [fontIcon]=\"child.icon\"></mat-icon>\n    <span>{{ child.label }}</span>\n  </button>\n</mat-menu>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["ruf-menu-label{display:block!important;min-width:100%}ruf-menu-label ruf-labeled-icon{-webkit-box-pack:center;justify-content:center;width:100%}ruf-menu-label ruf-labeled-icon>span{-webkit-box-flex:0;flex:0 1 auto}.ruf-menu-label-wrap{position:relative}.ruf-menu-label-trigger{bottom:0;left:0;position:absolute;right:0;top:0}.ruf-menu-label-chevron{color:currentColor;margin-left:5px}.ruf-menu-label-container{-webkit-box-align:center;align-items:center;display:-webkit-inline-box;display:inline-flex}.ruf-cust-menu .mat-menu-content{border-style:solid;border-width:0 1px 1px;min-width:7.5rem;padding-bottom:0;padding-top:0}.ruf-cust-menu .mat-menu-content .mat-menu-item{height:2.75rem;line-height:2.75rem}.ruf-cust-menu .mat-menu-content .ruf-menu-selected{text-decoration:underline}"]
        })
    ], RufMenuLabelComponent);
    return RufMenuLabelComponent;
}());
export { RufMenuLabelComponent };
