// @ruf/shell/esm5/src/dialog/dialog-panel.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, ChangeDetectionStrategy, Component, Directive, HostBinding, ViewEncapsulation, Input, ContentChild, } from '@angular/core';
import { RufShellIntl } from '../utils/shell-intl';
var RufDialogHeaderDirective = /** @class */ (function () {
    function RufDialogHeaderDirective() {
    }
    RufDialogHeaderDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-dialog-header' // tslint:disable-line
        })
    ], RufDialogHeaderDirective);
    return RufDialogHeaderDirective;
}());
export { RufDialogHeaderDirective };
var RufDialogContentDirective = /** @class */ (function () {
    function RufDialogContentDirective() {
    }
    RufDialogContentDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-dialog-content' // tslint:disable-line
        })
    ], RufDialogContentDirective);
    return RufDialogContentDirective;
}());
export { RufDialogContentDirective };
var RufDialogFooterDirective = /** @class */ (function () {
    function RufDialogFooterDirective() {
    }
    RufDialogFooterDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-dialog-footer' // tslint:disable-line
        })
    ], RufDialogFooterDirective);
    return RufDialogFooterDirective;
}());
export { RufDialogFooterDirective };
var RufDialogPanelComponent = /** @class */ (function () {
    function RufDialogPanelComponent(_intl, changeDetectorRef) {
        this._intl = _intl;
        this.tabindex = 0;
        this.separator = false;
        this._intlChanges = _intl.changes.subscribe(function () { return changeDetectorRef.markForCheck(); });
    }
    Object.defineProperty(RufDialogPanelComponent.prototype, "closeIconTitle", {
        get: function () {
            return this._closeIconTitle || this._intl.dialog.closeIconTitle;
        },
        set: function (value) {
            this._closeIconTitle = value;
        },
        enumerable: true,
        configurable: true
    });
    RufDialogPanelComponent.prototype.ngOnDestroy = function () {
        this._intlChanges.unsubscribe();
    };
    tslib_1.__decorate([
        ContentChild(RufDialogFooterDirective, { static: false }),
        tslib_1.__metadata("design:type", RufDialogFooterDirective)
    ], RufDialogPanelComponent.prototype, "footer", void 0);
    tslib_1.__decorate([
        HostBinding('attr.tabindex'),
        tslib_1.__metadata("design:type", Object)
    ], RufDialogPanelComponent.prototype, "tabindex", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufDialogPanelComponent.prototype, "separator", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufDialogPanelComponent.prototype, "closeIconTitle", null);
    RufDialogPanelComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-dialog-panel',
            template: "<div class=\"ruf-dialog-panel\" rufId=\"dialog_panel\" [ngClass]=\"{'ruf-dialog-separator-enabled': separator}\">\n  <div #title mat-dialog-title class=\"ruf-dialog-header\" rufId=\"dialog_header\">\n    <div class=\"ruf-dialog-header-bar\">\n      <ng-content select=\"ruf-dialog-header\"></ng-content>\n      <span class=\"ruf-dialog-header-filler\"></span>\n      <button type=\"button\" mat-icon-button fisStyle mat-dialog-close rufId=\"dialog_close_icon\" [title]=\"closeIconTitle\">\n        <mat-icon fisIcon=\"close\" rufIconStyle=\"sm\" alt=\"fis-close-icon\"></mat-icon>\n      </button>\n    </div>\n    <hr *ngIf=\"title.children.length > 0 && separator\" class=\"ruf-dialog-separator\">\n  </div>\n  <div mat-dialog-content class=\"ruf-dialog-content\" rufId=\"dialog_content\">\n    <ng-content select=\"ruf-dialog-content\"></ng-content>\n  </div>\n  <div #actions mat-dialog-actions class=\"ruf-dialog-footer\" [class.ruf-footer-empty]=\"!footer\" rufId=\"dialog_footer\">\n    <hr *ngIf=\"actions.children.length > 0  && separator\" class=\"ruf-dialog-separator\">\n    <ng-content select=\"ruf-dialog-footer\"></ng-content>\n  </div>\n  <ng-content></ng-content>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ruf-dialog-header{margin-bottom:1.125rem;min-height:2.25rem;position:relative}.ruf-dialog-header .ruf-dialog-header-bar{display:block}.ruf-dialog-header .ruf-dialog-header-bar ruf-dialog-header{display:block;padding-right:.9375rem;width:90%}.ruf-dialog-header .ruf-dialog-header-bar .mat-icon-button{position:absolute;right:0;top:-.1875rem}.mat-dialog-container,.ruf-dialog-panel{padding:1.25rem}mat-dialog-container.mat-dialog-container{border-radius:4px}mat-dialog-container.mat-dialog-container .ruf-dialog-panel{margin:-1.25rem}mat-dialog-container.mat-dialog-container .ruf-dialog-panel .ruf-dialog-footer{display:block;margin-bottom:.625rem;margin-top:1.25rem;min-height:2.625rem;padding:0}mat-dialog-container.mat-dialog-container .ruf-dialog-panel .ruf-dialog-footer ruf-dialog-footer{display:block;padding-top:.5rem}mat-dialog-container.mat-dialog-container .ruf-dialog-panel .ruf-dialog-footer.ruf-footer-empty{display:none}mat-dialog-container.mat-dialog-container .ruf-dialog-separator{margin-left:-1.25rem;margin-right:-1.25rem}mat-dialog-container.mat-dialog-container .ruf-dialog-content{margin:-1.25rem;padding:.625rem 1.25rem}.ruf-dialog-separator{border-style:solid}.ruf-dialog-footer>.ruf-dialog-separator{margin-bottom:0;margin-top:.5rem}mat-dialog-container.mat-dialog-container .ruf-dialog-separator-enabled .ruf-dialog-header{min-height:2.9375rem}mat-dialog-container.mat-dialog-container .ruf-dialog-separator-enabled.ruf-dialog-panel .ruf-dialog-footer{min-height:3.3125rem}mat-dialog-container.mat-dialog-container .ruf-dialog-separator-enabled.ruf-dialog-panel .ruf-dialog-footer ruf-dialog-footer{padding-top:1.125rem}mat-dialog-container.mat-dialog-container .ruf-dialog-separator-enabled .ruf-dialog-content{padding:1.25rem}"]
        }),
        tslib_1.__metadata("design:paramtypes", [RufShellIntl, ChangeDetectorRef])
    ], RufDialogPanelComponent);
    return RufDialogPanelComponent;
}());
export { RufDialogPanelComponent };
