// @ruf/shell/esm5/src/sitemap/sitemap.component.js
import * as tslib_1 from "tslib";
import { Input, Output, EventEmitter, Component, ViewEncapsulation, ChangeDetectionStrategy, ElementRef, QueryList, HostBinding, ViewChildren } from '@angular/core';
import { Observable } from 'rxjs';
import { RufKeyboardNavigation, RufKeyCodes } from '../input/keyboard';
import { RufFocusableDirective } from '../utils/focusable';
var RufSitemapComponent = /** @class */ (function () {
    function RufSitemapComponent(_elementRef) {
        this._elementRef = _elementRef;
        this.role = 'menubar';
        this.select = new EventEmitter();
        this.direction = 'row';
    }
    RufSitemapComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (this.sitemapItems.length > 0) {
            this.addNewKeyboardNav();
        }
        this.sitemapItems.changes.subscribe(function (changes) {
            if (_this.sitemapItems.length > 0) {
                _this.addNewKeyboardNav();
            }
        });
    };
    RufSitemapComponent.prototype.addNewKeyboardNav = function () {
        this._keyboardNav = new RufKeyboardNavigation();
        this._keyboardNav.setDirection(this.direction);
        this._keyboardNav.init(this._elementRef, this.sitemapItems);
        this._keyboardNav.addTabIndex(this.sitemapItems);
    };
    // emit the currently selected
    RufSitemapComponent.prototype.selectItem = function (sitemapItem) {
        this.select.emit(sitemapItem);
    };
    RufSitemapComponent.prototype.onKey = function (event, item) {
        this._keyboardNav.onKeyDown(event);
        switch (event.keyCode) {
            case RufKeyCodes.ENTER:
            case RufKeyCodes.SPACE:
                this.selectItem(item);
                break;
        }
    };
    RufSitemapComponent.prototype.uniqueItem = function (index, item) {
        return item.path;
    };
    tslib_1.__decorate([
        HostBinding('attr.role'),
        tslib_1.__metadata("design:type", Object)
    ], RufSitemapComponent.prototype, "role", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Observable)
    ], RufSitemapComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], RufSitemapComponent.prototype, "select", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufSitemapComponent.prototype, "direction", void 0);
    tslib_1.__decorate([
        ViewChildren(RufFocusableDirective),
        tslib_1.__metadata("design:type", QueryList)
    ], RufSitemapComponent.prototype, "sitemapItems", void 0);
    RufSitemapComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-sitemap',
            template: "<div class=\"ruf-sitemap-wrap\">\n  <div *ngFor=\"let parent of items | async; trackBy: uniqueItem \" class=\"ruf-sitemap-parent\">\n    <small *ngIf=\"parent.context\" class=\"ruf-sitemap-context\">{{parent.context}}\n      <span class=\"fis-icon-chevron-double\"></span>\n    </small>\n    <div class=\"ruf-title\">{{ parent.label }}</div><br>\n\n    <ul class=\"ruf-sitemap-children\">\n      <li *ngFor=\"let child of parent.children; trackBy: uniqueItem\" class=\"ruf-sitemap-child\">\n        <a (click)=\"selectItem(child);\" (keydown)=\"onKey($event, child)\" role=\"menuitem\" title=\"{{ child.label }}\" rufFocusable>{{ child.label }}</a>\n      </li>\n    </ul>\n  </div>\n</div>\n\n",
            encapsulation: ViewEncapsulation.None,
            providers: [],
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["ruf-sitemap{display:block;max-height:100%;overflow-y:auto;padding:20px;width:100%}ruf-sitemap .ruf-sitemap-wrap{display:-webkit-box;display:flex;-webkit-box-flex:1;flex:1 1 auto;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-flow:row wrap}ruf-sitemap .fis-icon-chevron-double{-webkit-transform:rotate(-90deg);transform:rotate(-90deg)}ruf-sitemap .ruf-sitemap-context{display:block;padding-left:10px}ruf-sitemap .ruf-sitemap-parent{width:200px}ruf-sitemap .ruf-sitemap-parent .ruf-title{cursor:default;display:inline-block;padding:5px 10px;text-transform:uppercase;white-space:pre-wrap;word-wrap:break-word}ruf-sitemap .ruf-sitemap-parent ul{margin:4px 4px 16px 0;padding-left:25px}ruf-sitemap .ruf-sitemap-parent ul .ruf-title{list-style-type:none}ruf-sitemap .ruf-sitemap-parent ul li{padding:4px}ruf-sitemap .ruf-sitemap-parent ul li a{cursor:pointer}ruf-sitemap .ruf-sitemap-parent ul li a:hover{text-decoration:underline}ruf-sitemap::-webkit-scrollbar{width:.5em}ruf-sitemap::-webkit-scrollbar-button{height:.5em}"]
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef])
    ], RufSitemapComponent);
    return RufSitemapComponent;
}());
export { RufSitemapComponent };
