// @ruf/store-packager/lib/generate-package-metadata.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const project_scanner_1 = require("@ruf/project-scanner");
const bundler_1 = require("@ruf/bundler");
const _ = require("lodash");
function generatePackageMetadata(options) {
    return __awaiter(this, void 0, void 0, function* () {
        console.log('RUF Store Packager Invoked!!!', options.flags);
        const storeMetadata = {
            htmldocs: '',
            contents: [],
            examples: [],
            logo: '',
            screenshots: '',
            sections: ''
        };
        const rootDir = path_1.dirname(options.flags.root);
        const examples = yield extractExamples(rootDir);
        storeMetadata.examples = examples;
        const readmes = yield extractReadmes(rootDir, rootDir);
        storeMetadata.readmes = readmes;
        const expaths = examples.map((ex) => Object.keys(ex));
        const readmePaths = Object.keys(readmes);
        project_scanner_1.traverseAngularProgram([path_1.join(process.cwd(), options.flags.root)], (entry) => {
            const compEntry = populateMetadata(entry, rootDir);
            if (compEntry.type === 'Directive' ||
                compEntry.type === 'Component' ||
                compEntry.type === 'Pipe') {
                compEntry.examples = getCompPaths(expaths && expaths.length > 0 ? expaths[0] : [], `${path_1.dirname(compEntry.path)}/examples`);
            }
            compEntry.readme = getCompPaths(readmePaths, path_1.dirname(compEntry.path))[0];
            storeMetadata.contents.push(compEntry);
        });
        return storeMetadata;
    });
}
exports.generatePackageMetadata = generatePackageMetadata;
function getCompPaths(paths, compDir) {
    const filteredPaths = paths.filter(eachPath => path_1.dirname(eachPath) === compDir);
    if (filteredPaths && filteredPaths.length > 0) {
        return filteredPaths;
    }
    return [];
}
function populateMetadata(entry, projectRoot) {
    return {
        type: entry.type,
        name: entry.name,
        path: getRelativePath(entry.filePath, path_1.join(process.cwd(), projectRoot).replace(/\\/g, '/')),
        jsDoc: entry.jsDoc,
        settings: entry.settings
    };
}
function getRelativePath(fromPath = '', toPath = '') {
    return fromPath.substr(toPath.length);
}
function extractExamples(projectRoot) {
    return __awaiter(this, void 0, void 0, function* () {
        const resultObj = [];
        const exObj = {};
        yield bundler_1.findExamples({ cwd: projectRoot }, (ex, info) => {
            const fullPath = info.dir.replace(/\\/g, '/');
            const examplePath = fullPath.substr(projectRoot.length);
            exObj[examplePath] = ex;
        });
        if (!_.isEmpty(exObj)) {
            resultObj.push(exObj);
        }
        return resultObj;
    });
}
function extractReadmes(pathStr, projectRoot) {
    return __awaiter(this, void 0, void 0, function* () {
        const resultObj = {};
        yield bundler_1.findReadmes({ cwd: pathStr }, (readmeFile) => {
            const readmePath = readmeFile.prop
                .replace(/\\/g, '/')
                .substr(projectRoot.length);
            resultObj[readmePath] = readmeFile.data;
        });
        return resultObj;
    });
}
