// @ruf/shell/esm5/src/inkbar/inkbar.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, HostBinding, Renderer2, ElementRef } from '@angular/core';
var RufInkbarDirective = /** @class */ (function () {
    function RufInkbarDirective(renderer, elm) {
        this.renderer = renderer;
        this.elm = elm;
        this._color = 'primary';
        this._position = 'bottom';
        this.rufInkbar = 'ruf-inkbar';
        // add default classes;
        this._updateClass(this._color, null);
        this._updateClass(this._position, null);
    }
    Object.defineProperty(RufInkbarDirective.prototype, "inkbarPosition", {
        get: function () {
            return this._position;
        },
        set: function (value) {
            var position = value || 'bottom';
            this._updateClass(position, this._position);
            this._position = position;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufInkbarDirective.prototype, "inkbarColor", {
        get: function () {
            return this._color;
        },
        set: function (value) {
            var color = value || 'primary';
            this._updateClass(color, this._color);
            this._color = color;
        },
        enumerable: true,
        configurable: true
    });
    // method to update the class based on the input
    RufInkbarDirective.prototype._updateClass = function (newValue, oldValue) {
        if (oldValue) {
            // remove class only if newValue is not null
            this.renderer.removeClass(this.elm.nativeElement, "ruf-inkbar-" + oldValue);
        }
        if (newValue) {
            this.renderer.addClass(this.elm.nativeElement, "ruf-inkbar-" + newValue);
        }
    };
    tslib_1.__decorate([
        HostBinding('class.ruf-inkbar'),
        tslib_1.__metadata("design:type", Object)
    ], RufInkbarDirective.prototype, "rufInkbar", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufInkbarDirective.prototype, "inkbarPosition", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufInkbarDirective.prototype, "inkbarColor", null);
    RufInkbarDirective = tslib_1.__decorate([
        Directive({
            selector: '[rufInkbar]',
            exportAs: 'rufInkbar'
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2, ElementRef])
    ], RufInkbarDirective);
    return RufInkbarDirective;
}());
export { RufInkbarDirective };
var RufInkbarActiveDirective = /** @class */ (function () {
    function RufInkbarActiveDirective() {
        this.rufInkbarItem = 'ruf-inkbar-item';
    }
    Object.defineProperty(RufInkbarActiveDirective.prototype, "_inkbarActive", {
        get: function () {
            return this.rufInkbarActive;
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        HostBinding('class.ruf-inkbar-item'),
        tslib_1.__metadata("design:type", Object)
    ], RufInkbarActiveDirective.prototype, "rufInkbarItem", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-inkbar-active'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], RufInkbarActiveDirective.prototype, "_inkbarActive", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], RufInkbarActiveDirective.prototype, "rufInkbarActive", void 0);
    RufInkbarActiveDirective = tslib_1.__decorate([
        Directive({
            selector: '[rufInkbarActive]',
            exportAs: 'rufInkbarActive'
        })
    ], RufInkbarActiveDirective);
    return RufInkbarActiveDirective;
}());
export { RufInkbarActiveDirective };
