// @ruf/shell/esm2015/src/sidemenu/sidemenu-item.component.js
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Inject, forwardRef, ElementRef } from '@angular/core';
import { RufMenubarItemComponent } from '../menubar/menubar-item.component';
import { RufSidemenuComponent } from './sidemenu.component';
let RufSidemenuItemComponent = class RufSidemenuItemComponent extends RufMenubarItemComponent {
    constructor(menubar, _elementRef) {
        super(menubar);
        this.menubar = menubar;
        this._elementRef = _elementRef;
    }
};
RufSidemenuItemComponent = tslib_1.__decorate([
    Component({
        selector: 'ruf-sidemenu-item, [rufSidemenuItem]',
        template: `
      <ng-content></ng-content>
  `,
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    tslib_1.__param(0, Inject(forwardRef(() => RufSidemenuComponent))),
    tslib_1.__metadata("design:paramtypes", [RufSidemenuComponent,
        ElementRef])
], RufSidemenuItemComponent);
export { RufSidemenuItemComponent };
