// @ruf/shell/esm2015/src/inkbar/inkbar.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, HostBinding, Renderer2, ElementRef } from '@angular/core';
let RufInkbarDirective = class RufInkbarDirective {
    constructor(renderer, elm) {
        this.renderer = renderer;
        this.elm = elm;
        this._color = 'primary';
        this._position = 'bottom';
        this.rufInkbar = 'ruf-inkbar';
        // add default classes;
        this._updateClass(this._color, null);
        this._updateClass(this._position, null);
    }
    get inkbarPosition() {
        return this._position;
    }
    set inkbarPosition(value) {
        const position = value || 'bottom';
        this._updateClass(position, this._position);
        this._position = position;
    }
    get inkbarColor() {
        return this._color;
    }
    set inkbarColor(value) {
        const color = value || 'primary';
        this._updateClass(color, this._color);
        this._color = color;
    }
    // method to update the class based on the input
    _updateClass(newValue, oldValue) {
        if (oldValue) {
            // remove class only if newValue is not null
            this.renderer.removeClass(this.elm.nativeElement, `ruf-inkbar-${oldValue}`);
        }
        if (newValue) {
            this.renderer.addClass(this.elm.nativeElement, `ruf-inkbar-${newValue}`);
        }
    }
};
tslib_1.__decorate([
    HostBinding('class.ruf-inkbar'),
    tslib_1.__metadata("design:type", Object)
], RufInkbarDirective.prototype, "rufInkbar", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufInkbarDirective.prototype, "inkbarPosition", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufInkbarDirective.prototype, "inkbarColor", null);
RufInkbarDirective = tslib_1.__decorate([
    Directive({
        selector: '[rufInkbar]',
        exportAs: 'rufInkbar'
    }),
    tslib_1.__metadata("design:paramtypes", [Renderer2, ElementRef])
], RufInkbarDirective);
export { RufInkbarDirective };
let RufInkbarActiveDirective = class RufInkbarActiveDirective {
    constructor() {
        this.rufInkbarItem = 'ruf-inkbar-item';
    }
    get _inkbarActive() {
        return this.rufInkbarActive;
    }
};
tslib_1.__decorate([
    HostBinding('class.ruf-inkbar-item'),
    tslib_1.__metadata("design:type", Object)
], RufInkbarActiveDirective.prototype, "rufInkbarItem", void 0);
tslib_1.__decorate([
    HostBinding('class.ruf-inkbar-active'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], RufInkbarActiveDirective.prototype, "_inkbarActive", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], RufInkbarActiveDirective.prototype, "rufInkbarActive", void 0);
RufInkbarActiveDirective = tslib_1.__decorate([
    Directive({
        selector: '[rufInkbarActive]',
        exportAs: 'rufInkbarActive'
    })
], RufInkbarActiveDirective);
export { RufInkbarActiveDirective };
