// @ruf/shell/esm5/src/toolbar/toolbar.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input, ViewEncapsulation, ElementRef, ChangeDetectionStrategy, Directive, ContentChildren, QueryList, HostListener } from '@angular/core';
import { RufKeyboardNavigation, RufKeyCodes } from '../input/keyboard';
import { RufFocusableDirective } from '../utils/focusable';
import { RufShellIntl } from '../utils/shell-intl';
import { RufMixinBase } from '../common/color/color';
var RufToolbarRowDirective = /** @class */ (function () {
    function RufToolbarRowDirective() {
    }
    RufToolbarRowDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-toolbar-row' // tslint:disable-line
        })
    ], RufToolbarRowDirective);
    return RufToolbarRowDirective;
}());
export { RufToolbarRowDirective };
var RufToolbarComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RufToolbarComponent, _super);
    function RufToolbarComponent(_elementRef, _intl, changeDetectorRef) {
        var _this = _super.call(this, _elementRef) || this;
        _this._elementRef = _elementRef;
        _this._intl = _intl;
        _this.children = [];
        _this._intlChanges = _intl.changes.subscribe(function () {
            changeDetectorRef.markForCheck();
        });
        return _this;
    }
    Object.defineProperty(RufToolbarComponent.prototype, "ariaLabel", {
        get: function () {
            return this._ariaLabel || this._intl.toolbar.aria.label;
        },
        /* tslint:disable:no-input-rename */
        set: function (label) {
            this._ariaLabel = label;
        },
        enumerable: true,
        configurable: true
    });
    RufToolbarComponent.prototype.ngAfterContentInit = function () {
        this._keyboardNav = new RufKeyboardNavigation();
        this._keyboardNav.init(this._elementRef, this.toolbarItems);
        this._keyboardNav.addTabIndex(this.toolbarItems);
    };
    RufToolbarComponent.prototype.onKey = function (event) {
        switch (event.keyCode) {
            case RufKeyCodes.SPACE:
            case RufKeyCodes.ENTER:
                event.stopPropagation();
                break;
            default:
                this._keyboardNav.onKeyDown(event);
        }
    };
    RufToolbarComponent.prototype.ngOnDestroy = function () {
        this._intlChanges.unsubscribe();
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], RufToolbarComponent.prototype, "color", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], RufToolbarComponent.prototype, "ariaLabel", null);
    tslib_1.__decorate([
        ContentChildren(RufFocusableDirective, { descendants: true }),
        tslib_1.__metadata("design:type", QueryList)
    ], RufToolbarComponent.prototype, "toolbarItems", void 0);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [KeyboardEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufToolbarComponent.prototype, "onKey", null);
    RufToolbarComponent = tslib_1.__decorate([
        Component({
            selector: 'ruf-toolbar',
            template: "<div class=\"ruf-toolbar-rows\" rufId=\"toolbar\" >\n  <ruf-toolbar-row  class=\"ruf-toolbar-first-row\" rufId=\"toolbar_row\" role=\"toolbar\" [attr.aria-label]=\"ariaLabel\">\n    <ng-content></ng-content>\n  </ruf-toolbar-row>\n  <ng-content select=\"ruf-toolbar-row\"></ng-content>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["ruf-toolbar{display:block;flex-shrink:0;min-height:inherit;overflow:hidden}ruf-toolbar .mat-select{padding-top:0}.ruf-toolbar-separator{border-left-style:solid;border-left-width:1px;display:inline-block;height:1.4375rem;margin:0 .625rem;vertical-align:middle;word-wrap:nowrap}.ruf-toolbar-rows{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;min-height:inherit}ruf-toolbar-row{-webkit-box-align:center;align-items:center;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;min-height:inherit}ruf-toolbar-row>*{margin-left:.625rem}ruf-toolbar-row>:nth-last-child(1){margin-left:.625rem;margin-right:.625rem}ruf-toolbar-row>:nth-last-child(1).ruf-square{margin-right:0}ruf-toolbar-row>:only-child{border:0}"]
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef,
            RufShellIntl,
            ChangeDetectorRef])
    ], RufToolbarComponent);
    return RufToolbarComponent;
}(RufMixinBase));
export { RufToolbarComponent };
