// @ruf/shell/esm2015/src/trigger/trigger.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Renderer2, HostListener, Input } from '@angular/core';
import { CdkOverlayOrigin } from '@angular/cdk/overlay';
import { Subscription } from 'rxjs';
import { throwTargetMissingError } from './trigger.interface';
let RufTriggerDirective = class RufTriggerDirective extends CdkOverlayOrigin {
    constructor(elRef, renderer) {
        super(elRef);
        this.elRef = elRef;
        this.renderer = renderer;
        this.subscription = Subscription.EMPTY;
    }
    ngAfterViewInit() {
        this._checkTarget();
        if (this.target.hasPopup()) {
            this.renderer.setAttribute(this.elRef.nativeElement, 'aria-haspopup', 'true');
            this.renderer.setAttribute(this.elRef.nativeElement, 'aria-expanded', 'false');
        }
        else {
            this.renderer.removeAttribute(this.elRef.nativeElement, 'aria-haspopup');
            this.renderer.removeAttribute(this.elRef.nativeElement, 'aria-expanded');
        }
        if (this.subscription === Subscription.EMPTY) {
            this.subscription = this.target.toggle.subscribe((opened) => {
                this.renderer.setAttribute(this.elRef.nativeElement, 'aria-expanded', opened ? 'true' : 'false');
                if (!opened) {
                    this.elRef.nativeElement.focus();
                }
            });
        }
    }
    handleClickEvent(event) {
        this.target.handleClick(event);
    }
    handleKeydownEvent(event) {
        this.target.handleKeydown(event);
    }
    _checkTarget() {
        if (!this.target) {
            throwTargetMissingError();
        }
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
};
tslib_1.__decorate([
    Input('rufTrigger'),
    tslib_1.__metadata("design:type", Object)
], RufTriggerDirective.prototype, "target", void 0);
tslib_1.__decorate([
    HostListener('click', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], RufTriggerDirective.prototype, "handleClickEvent", null);
tslib_1.__decorate([
    HostListener('keydown', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], RufTriggerDirective.prototype, "handleKeydownEvent", null);
RufTriggerDirective = tslib_1.__decorate([
    Directive({
        selector: '[rufTrigger]',
        exportAs: 'rufTrigger'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], RufTriggerDirective);
export { RufTriggerDirective };
