// @ruf/shell/esm5/src/toggle/toggle.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostBinding, HostListener, Input, Output, EventEmitter } from '@angular/core';
import { ENTER, SPACE } from '@angular/cdk/keycodes';
import { CdkOverlayOrigin } from '@angular/cdk/overlay';
/**
 * Directive that can be applied on any element, and that toggles an attribute that can be used by the underlying element as need be.
 */
var RufToggleDirective = /** @class */ (function (_super) {
    tslib_1.__extends(RufToggleDirective, _super);
    function RufToggleDirective(elRef) {
        var _this = _super.call(this, elRef) || this;
        _this.elRef = elRef;
        // boolean property that gets toggled when the underlying element is clicked
        // tslint:disable-next-line:no-input-rename
        _this.value = false;
        _this.chevron = false;
        _this.arrow = false;
        _this.toggleChange = new EventEmitter();
        _this.toggle = true;
        return _this;
    }
    Object.defineProperty(RufToggleDirective.prototype, "active", {
        get: function () {
            return this.value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufToggleDirective.prototype, "rufToggle", {
        set: function (val) {
            this.value = val || false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufToggleDirective.prototype, "rufToggleChevron", {
        set: function (val) {
            this.arrow = false;
            this.chevron = true;
            this.value = val || false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufToggleDirective.prototype, "rufToggleArrow", {
        set: function (val) {
            this.arrow = true;
            this.chevron = false;
            this.value = val || false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufToggleDirective.prototype, "normal", {
        get: function () {
            return !this.chevron && !this.arrow;
        },
        enumerable: true,
        configurable: true
    });
    // select should focus the element and toggle value
    RufToggleDirective.prototype.select = function (event) {
        this.elRef.nativeElement.focus();
        this.toggleValue(event);
    };
    RufToggleDirective.prototype.toggleValue = function (event) {
        event.stopPropagation();
        event.preventDefault();
        // For Safari browser, event.stopPropagation() is not setting the value of
        // event.cancelBubble as true. So, we need to explicitly set it.
        event.cancelBubble = true;
        this.value = !this.value;
        this.toggleChange.emit({ trigger: this.elRef, value: this.value, event: event });
    };
    RufToggleDirective.prototype.onClick = function (event) {
        this.toggleValue(event);
    };
    RufToggleDirective.prototype.onKey = function (event) {
        switch (event.keyCode) {
            case ENTER:
            case SPACE:
                this.toggleValue(event);
                break;
            default:
        }
    };
    tslib_1.__decorate([
        HostBinding('class.ruf-toggle-active'),
        tslib_1.__metadata("design:type", Object)
    ], RufToggleDirective.prototype, "value", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-toggle-chevron'),
        tslib_1.__metadata("design:type", Object)
    ], RufToggleDirective.prototype, "chevron", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-toggle-arrow'),
        tslib_1.__metadata("design:type", Object)
    ], RufToggleDirective.prototype, "arrow", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-active'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], RufToggleDirective.prototype, "active", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], RufToggleDirective.prototype, "rufToggle", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], RufToggleDirective.prototype, "rufToggleChevron", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], RufToggleDirective.prototype, "rufToggleArrow", null);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], RufToggleDirective.prototype, "toggleChange", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-toggle'),
        tslib_1.__metadata("design:type", Object)
    ], RufToggleDirective.prototype, "toggle", void 0);
    tslib_1.__decorate([
        HostBinding('class.ruf-toggle-normal'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], RufToggleDirective.prototype, "normal", null);
    tslib_1.__decorate([
        HostListener('click', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufToggleDirective.prototype, "onClick", null);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [KeyboardEvent]),
        tslib_1.__metadata("design:returntype", void 0)
    ], RufToggleDirective.prototype, "onKey", null);
    RufToggleDirective = tslib_1.__decorate([
        Directive({
            selector: '[rufToggle], [rufToggleChevron], [rufToggleArrow]',
            exportAs: 'rufToggle'
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef])
    ], RufToggleDirective);
    return RufToggleDirective;
}(CdkOverlayOrigin));
export { RufToggleDirective };
