// @ruf/project-scanner/lib/angular-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const ts = require("typescript");
const path_1 = require("path");
const angularClassDecorators = [
    'Injectable',
    'Directive',
    'Pipe',
    'NgModule',
    'Component'
];
function getCompilerOptions(opts) {
    if (typeof opts === 'string') {
        const tsconf = JSON.parse(fs_1.readFileSync(opts).toString()).compilerOptions;
        tsconf.target = ts.ScriptTarget.ES5;
        tsconf.module = ts.ModuleKind.CommonJS;
        tsconf.moduleResolution = ts.ModuleResolutionKind.NodeJs;
        return tsconf;
    }
    return opts;
}
function traverseAngularProgram(fileNames, visitor, options = {
    compiler: {
        target: ts.ScriptTarget.ES5,
        module: ts.ModuleKind.CommonJS
    }
}) {
    const compilerOpts = getCompilerOptions(options.compiler);
    const program = ts.createProgram(fileNames, compilerOpts);
    const checker = program.getTypeChecker();
    for (const sourceFile of program.getSourceFiles()) {
        if (!sourceFile.isDeclarationFile) {
            ts.forEachChild(sourceFile, visit);
        }
    }
    return;
    function isNodeExported(node) {
        return ((ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export) !== 0 ||
            (!!node.parent && node.parent.kind === ts.SyntaxKind.SourceFile));
    }
    function visit(node) {
        if (!isNodeExported(node)) {
            return;
        }
        if (ts.isClassDeclaration(node) && node.name && node.decorators) {
            const symbol = checker.getSymbolAtLocation(node.name);
            const type = toAngularType(node.decorators);
            if (symbol && type) {
                visitor(buildEntry(node, symbol, type));
            }
        }
        else if (ts.isModuleDeclaration(node)) {
            ts.forEachChild(node, visit);
        }
    }
    function buildEntry(node, symbol, type) {
        const entry = Object.assign({ name: symbol.name, filePath: node.getSourceFile().fileName, dir: path_1.dirname(node.getSourceFile().fileName), jsDoc: ts.displayPartsToString(symbol.getDocumentationComment(checker)) }, type);
        return entry;
    }
    function propsToSettings(props) {
        return props.reduce((acc, prop) => {
            const init = prop.initializer;
            const name = prop.name.escapedText;
            if (!name.startsWith('style') &&
                !name.startsWith('template') &&
                !name.startsWith('animations')) {
                const val = (init.text && init.text.replace(/\s/g, '').replace(/,/g, ', ')) ||
                    (init.elements && init.elements.map((e) => e.getText()));
                if (val) {
                    acc[name] = val;
                }
            }
            return acc;
        }, {});
    }
    function toAngularType(decorators) {
        return decorators.reduce((acc, decorator) => {
            const symbol = checker.getSymbolAtLocation(decorator.expression.getFirstToken());
            if (symbol && angularClassDecorators.indexOf(symbol.name) !== -1) {
                const expr = decorator.expression;
                if (expr.arguments.length && expr.arguments[0].properties) {
                    const props = expr.arguments[0].properties;
                    return {
                        type: symbol.name,
                        settings: propsToSettings(props)
                    };
                }
                return {
                    type: symbol.name,
                    settings: {}
                };
            }
            return null;
        }, null);
    }
}
exports.traverseAngularProgram = traverseAngularProgram;
