// @ruf/shell/esm5/src/router/mdi-strategy.js
var RufMDIStrategy = /** @class */ (function () {
    function RufMDIStrategy(util) {
        this.util = util;
        this.handlers = {};
    }
    /**
     * Decides when the route should be stored
     * @param route the route that the user is currently on
     * @returns boolean indicating whether the route should or should not be stored
     */
    RufMDIStrategy.prototype.shouldDetach = function (route) {
        return true;
    };
    /**
     * Stores the complete route path and its component state (as ComponentRef) for later attachment
     * @param route This is stored for later comparison to requested routes, see `this.shouldAttach`
     * @param handle A DetachedRouteHandle which is later retrieved by this.retrieve to restore state
     */
    RufMDIStrategy.prototype.store = function (route, handle) {
        var path = this.util.getLeafNodePath(route);
        // path is '' when route has loadChildren.
        if (path) {
            if (route.component) {
                this.handlers[path] = handle;
            }
            // route does not have component. this means it is a parent route like ruf/shell/banner or ruf/shell/layout.
            // In this case, previously stored leaf node handler should be re-stored.
            if (!route.component && handle) {
                // This condition is when parent route is being stored but it's the child route whose reference needs to be stored instead.
                // In this case, retrieve the component ref from RufRouterUtilService and create a new DetachedRouteHandle.
                this.handlers[path] = {
                    componentRef: this.util.getComponentRef(path),
                    route: this.util.getLeafNode(handle['route'])
                };
            }
        }
    };
    /**
     * Determines whether there is a stored route and, if yes, should it be rendered in place of requested route
     * @param route The route the user requested
     * @returns boolean indicating whether or not to render the stored route
     */
    RufMDIStrategy.prototype.shouldAttach = function (route) {
        var path = this.util.getLeafNodePath(route);
        return !!route.routeConfig && !!this.handlers[path];
    };
    /**
     * Finds the locally stored instance of the requested route and returns it if it exists
     * @param route New route the user has requested
     * @returns DetachedRouteHandle object which can be used to render the component
     */
    RufMDIStrategy.prototype.retrieve = function (route) {
        if (!route.routeConfig) {
            return null;
        }
        if (route.routeConfig.loadChildren) {
            return null;
        }
        return this.handlers[this.util.fullPath(route)];
    };
    /**
     * Determines whether or not the current route should be reused
     * @param future The route the user is navigating to as triggered by the router
     * @param curr The route the user is currently on
     * @returns boolean indicating true if the user intends to leave the current route
     */
    RufMDIStrategy.prototype.shouldReuseRoute = function (future, curr) {
        return ((future.routeConfig === curr.routeConfig) &&
            (JSON.stringify(future.params) === JSON.stringify(curr.params)));
    };
    /**
     *  Remove component reference from handlers.
     *  Destroy the route component
     *  @param path The key against which DetachedRouteHandle with component reference is stored for detached route
     */
    RufMDIStrategy.prototype.destroy = function (path) {
        if (this.handlers[path] && this.util.getComponentRef(path)) {
            this.util.getComponentRef(path).destroy();
            delete this.handlers[path];
        }
    };
    return RufMDIStrategy;
}());
export { RufMDIStrategy };
