// @ruf/shell/esm5/src/router/outlet.directive.js
import * as tslib_1 from "tslib";
import { Attribute, ComponentFactoryResolver, ChangeDetectorRef, Directive, ViewContainerRef } from '@angular/core';
import { RouterOutlet, ChildrenOutletContexts, PRIMARY_OUTLET } from '@angular/router';
import { RufRouterUtilService } from './router-util.service';
var RufRouterOutletDirective = /** @class */ (function (_super) {
    tslib_1.__extends(RufRouterOutletDirective, _super);
    function RufRouterOutletDirective(rufParentContexts, rufLocation, rufResolver, rufName, rufChangeDetector, util) {
        var _this = _super.call(this, rufParentContexts, rufLocation, rufResolver, rufName, rufChangeDetector) || this;
        _this.rufParentContexts = rufParentContexts;
        _this.rufLocation = rufLocation;
        _this.rufResolver = rufResolver;
        _this.rufName = rufName;
        _this.rufChangeDetector = rufChangeDetector;
        _this.util = util;
        _this.rufName = rufName || PRIMARY_OUTLET;
        return _this;
    }
    RufRouterOutletDirective.prototype.detach = function () {
        var componentRef = _super.prototype.detach.call(this);
        var lastActivatedOutlet = this.util.lastActivatedOutlet;
        // Detach child routes when a parent route like /banner or /layout is being detached.
        if (lastActivatedOutlet && lastActivatedOutlet.isActivated) {
            var lastCompRef = lastActivatedOutlet.detach();
            this.util.setComponentRef(this.util.lastActivatedPath, lastCompRef);
        }
        else {
            this.util.setComponentRef(this.util.lastActivatedPath, componentRef);
        }
        return componentRef;
    };
    RufRouterOutletDirective.prototype.activateWith = function (activatedRoute, resolver) {
        this.storePreviousRoute(activatedRoute);
        _super.prototype.activateWith.call(this, activatedRoute, resolver);
    };
    RufRouterOutletDirective.prototype.attach = function (ref, activatedRoute) {
        this.storePreviousRoute(activatedRoute);
        _super.prototype.attach.call(this, ref, activatedRoute);
    };
    RufRouterOutletDirective.prototype.storePreviousRoute = function (activatedRoute) {
        this.util.lastActivatedPath = this.util.fullPath(activatedRoute.snapshot);
        if (!activatedRoute.children || activatedRoute.children.length === 0) {
            // Store router outlet only in case of leaf routes. This will be used when detaching parent routes.
            this.util.lastActivatedOutlet = this.rufParentContexts.getContext(this.rufName).outlet;
        }
    };
    RufRouterOutletDirective = tslib_1.__decorate([
        Directive({
            selector: 'ruf-router-outlet' // tslint:disable-line
        }),
        tslib_1.__param(3, Attribute('name')),
        tslib_1.__metadata("design:paramtypes", [ChildrenOutletContexts,
            ViewContainerRef,
            ComponentFactoryResolver, String, ChangeDetectorRef,
            RufRouterUtilService])
    ], RufRouterOutletDirective);
    return RufRouterOutletDirective;
}(RouterOutlet));
export { RufRouterOutletDirective };
