// @ruf/shell/esm5/src/router/router-util.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var RufRouterUtilService = /** @class */ (function () {
    function RufRouterUtilService() {
        this._routeComp = {};
    }
    Object.defineProperty(RufRouterUtilService.prototype, "lastActivatedPath", {
        get: function () {
            return this._lastActivatedPath;
        },
        set: function (path) {
            this._lastActivatedPath = path;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RufRouterUtilService.prototype, "lastActivatedOutlet", {
        get: function () {
            return this._lastActivatedOutlet;
        },
        set: function (outlet) {
            this._lastActivatedOutlet = outlet;
        },
        enumerable: true,
        configurable: true
    });
    RufRouterUtilService.prototype.setComponentRef = function (path, componentRef) {
        this._routeComp[path] = componentRef;
    };
    RufRouterUtilService.prototype.getComponentRef = function (path) {
        return this._routeComp[path];
    };
    RufRouterUtilService.prototype.fullPath = function (snapshot) {
        var path = '';
        var rootElems = snapshot.pathFromRoot;
        rootElems.forEach(function (child) {
            var urlArray = child.url;
            urlArray.forEach(function (url) {
                path += "/" + url.path;
            });
        });
        return path;
    };
    RufRouterUtilService.prototype.getLeafNodePath = function (route) {
        var path;
        // if (!route.routeConfig.loadChildren) {
        // If !route.component, this is a parent route. Return path of the leaf node in this case.
        if (!route.component) {
            var leafNode = this.getLeafNode(route);
            path = this.fullPath(leafNode);
        }
        else {
            path = this.fullPath(route);
        }
        // }
        return path;
    };
    RufRouterUtilService.prototype.getLeafNode = function (route) {
        if (!route.children || route.children.length === 0) {
            return route;
        }
        return this.getLeafNode(route.firstChild || route.children[0]);
    };
    RufRouterUtilService = tslib_1.__decorate([
        Injectable()
    ], RufRouterUtilService);
    return RufRouterUtilService;
}());
export { RufRouterUtilService };
