// @ruf/shell/esm2015/src/banner/banner.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input, ViewEncapsulation, Directive, ContentChild, ChangeDetectionStrategy } from '@angular/core';
import { RufShellIntl } from '../utils/shell-intl';
let RufBannerRowDirective = class RufBannerRowDirective {
};
RufBannerRowDirective = tslib_1.__decorate([
    Directive({
        selector: 'ruf-banner-row' // tslint:disable-line
    })
], RufBannerRowDirective);
export { RufBannerRowDirective };
let RufBannerControlsDirective = class RufBannerControlsDirective {
};
RufBannerControlsDirective = tslib_1.__decorate([
    Directive({
        selector: '[rufBannerControls]' // tslint:disable-line
    })
], RufBannerControlsDirective);
export { RufBannerControlsDirective };
let RufBannerBrandComponent = class RufBannerBrandComponent {
    constructor(_intl, changeDetectorRef) {
        this._intl = _intl;
        this.logoClass = 'fis-logo';
        this.showLogo = true;
        this._intlChanges = _intl.changes.subscribe(() => changeDetectorRef.markForCheck());
    }
    get altText() {
        return this._altText || this._intl.banner.altText;
    }
    set altText(value) {
        this._altText = value;
    }
    get ariaLabel() {
        return this._ariaLabel || this._intl.banner.aria.label;
    }
    set ariaLabel(value) {
        this._ariaLabel = value;
    }
    get ariaLabelledby() {
        return this._ariaLabelledby || this._intl.banner.aria.labelledBy;
    }
    set ariaLabelledby(value) {
        this._ariaLabel = value;
    }
    ngOnDestroy() {
        this._intlChanges.unsubscribe();
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], RufBannerBrandComponent.prototype, "logoUrl", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], RufBannerBrandComponent.prototype, "title", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufBannerBrandComponent.prototype, "logoClass", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], RufBannerBrandComponent.prototype, "showLogo", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufBannerBrandComponent.prototype, "altText", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufBannerBrandComponent.prototype, "ariaLabel", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], RufBannerBrandComponent.prototype, "ariaLabelledby", null);
RufBannerBrandComponent = tslib_1.__decorate([
    Component({
        selector: 'ruf-banner-brand',
        template: `
    <div rufId [class]="'ruf-banner-brand-logo ' + (!logoUrl && logoClass)" *ngIf="showLogo" >
      <span hidden="true" aria-hidden="false" [attr.aria-label]="ariaLabel"
      [attr.aria-labelledby]="ariaLabelledby">{{ariaLabel}}</span>
      <img rufId *ngIf="logoUrl" [src]="logoUrl" [attr.alt]="altText">
    </div>
    <div rufId class="ruf-toolbar-separator" *ngIf="title && showLogo"></div>
    <h1 *ngIf="title" class="ruf-banner-brand-title">{{title}}</h1>
  `,
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    tslib_1.__metadata("design:paramtypes", [RufShellIntl, ChangeDetectorRef])
], RufBannerBrandComponent);
export { RufBannerBrandComponent };
let RufBannerComponent = class RufBannerComponent {
    constructor() { }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], RufBannerComponent.prototype, "color", void 0);
tslib_1.__decorate([
    ContentChild(RufBannerControlsDirective, { static: false }),
    tslib_1.__metadata("design:type", RufBannerControlsDirective)
], RufBannerComponent.prototype, "controls", void 0);
RufBannerComponent = tslib_1.__decorate([
    Component({
        selector: 'ruf-banner',
        template: "<header class=\"ruf-banner-header\" role=\"banner\">\n  <ruf-toolbar [color]=\"color\" class=\"ruf-banner-toolbar\" rufId=\"banner_toolbar\">\n    <ng-content></ng-content>\n    <div *ngIf=\"controls\" class=\"ruf-banner-filler\"></div>\n    <ng-content select=\"[rufBannerControls]\"></ng-content>\n    <ng-content select=\"ruf-banner-row\" ngProjectAs=\"ruf-toolbar-row\"></ng-content>\n  </ruf-toolbar>\n</header>\n",
        encapsulation: ViewEncapsulation.None,
        styles: ["ruf-banner{display:block;flex-shrink:0;min-height:3rem;overflow:hidden;z-index:100}.ruf-banner-toolbar{height:inherit}.ruf-banner-toolbar ruf-banner-row,.ruf-banner-toolbar ruf-toolbar-row{margin:0 auto;padding:0;width:100%}.ruf-banner-header{min-height:inherit}ruf-banner-brand{-webkit-box-align:center;align-items:center;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;height:3rem}ruf-banner-brand .ruf-toolbar-separator{display:none}.ruf-banner-filler{-webkit-box-flex:1;flex:1 1 auto}.ruf-banner-brand-logo{line-height:0}.ruf-banner-brand-logo.fis-logo{background-size:4.375rem;height:2.375rem;max-width:4.375rem;width:4.375rem}.ruf-banner-brand-title{display:none;white-space:nowrap}@media (min-width:960px){.ruf-banner-brand-title,ruf-banner-brand .ruf-toolbar-separator{display:block}}"]
    }),
    tslib_1.__metadata("design:paramtypes", [])
], RufBannerComponent);
export { RufBannerComponent };
